{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c10ef9b7-2644-4e9e-b606-951a46c30741",
   "metadata": {},
   "source": [
    "# Background Information\n",
    "\n",
    "In this notebook I will set up a basic flow for uploading and dowloading files. Formating datasets and ultimately running cytetype."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d60fdade-301f-4be5-aede-eb66b7ed7e41",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Importing Require Libraries\n",
    "import cytetype\n",
    "import scanpy as sc\n",
    "import session_info\n",
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "from dotenv import load_dotenv\n",
    "import json\n",
    "\n",
    "load_dotenv()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9c67fbd3-b6f9-4ee9-8dcf-588479c9c267",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/session_info/main.py:213: UserWarning: The '__version__' attribute is deprecated and will be removed in MarkupSafe 3.1. Use feature detection, or `importlib.metadata.version(\"markupsafe\")`, instead.\n",
      "  mod_version = _find_version(mod.__version__)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<details>\n",
       "<summary>Click to view session information</summary>\n",
       "<pre>\n",
       "-----\n",
       "cytetype            0.4.0\n",
       "dotenv              NA\n",
       "numpy               2.2.0\n",
       "pandas              2.3.0\n",
       "scanpy              1.11.2\n",
       "session_info        v1.0.1\n",
       "-----\n",
       "</pre>\n",
       "<details>\n",
       "<summary>Click to view modules imported as dependencies</summary>\n",
       "<pre>\n",
       "PIL                         11.2.1\n",
       "anndata                     0.11.4\n",
       "annotated_types             0.7.0\n",
       "anyio                       NA\n",
       "appnope                     0.1.4\n",
       "arrow                       1.3.0\n",
       "asttokens                   NA\n",
       "attr                        25.3.0\n",
       "attrs                       25.3.0\n",
       "babel                       2.17.0\n",
       "certifi                     2025.04.26\n",
       "charset_normalizer          3.4.2\n",
       "comm                        0.2.2\n",
       "cycler                      0.12.1\n",
       "cython_runtime              NA\n",
       "dateutil                    2.9.0.post0\n",
       "debugpy                     1.8.14\n",
       "decorator                   5.2.1\n",
       "defusedxml                  0.7.1\n",
       "executing                   2.2.0\n",
       "fastjsonschema              NA\n",
       "fqdn                        NA\n",
       "h5py                        3.14.0\n",
       "idna                        3.10\n",
       "ipykernel                   6.29.5\n",
       "isoduration                 NA\n",
       "jedi                        0.19.2\n",
       "jinja2                      3.1.6\n",
       "joblib                      1.5.1\n",
       "json5                       0.12.0\n",
       "jsonpointer                 3.0.0\n",
       "jsonschema                  4.23.0\n",
       "jsonschema_specifications   NA\n",
       "jupyter_events              0.12.0\n",
       "jupyter_server              2.16.0\n",
       "jupyterlab_server           2.27.3\n",
       "kiwisolver                  1.4.8\n",
       "legacy_api_wrap             NA\n",
       "llvmlite                    0.44.0\n",
       "loguru                      0.7.3\n",
       "markupsafe                  3.0.2\n",
       "matplotlib                  3.10.3\n",
       "mpl_toolkits                NA\n",
       "natsort                     8.4.0\n",
       "nbformat                    5.10.4\n",
       "numba                       0.61.2\n",
       "overrides                   NA\n",
       "packaging                   25.0\n",
       "parso                       0.8.4\n",
       "platformdirs                4.3.8\n",
       "prometheus_client           NA\n",
       "prompt_toolkit              3.0.51\n",
       "psutil                      7.0.0\n",
       "pure_eval                   0.2.3\n",
       "pydantic                    2.11.5\n",
       "pydantic_core               2.33.2\n",
       "pydev_ipython               NA\n",
       "pydevconsole                NA\n",
       "pydevd                      3.2.3\n",
       "pydevd_file_utils           NA\n",
       "pydevd_plugins              NA\n",
       "pydevd_tracing              NA\n",
       "pygments                    2.19.1\n",
       "pyparsing                   3.2.3\n",
       "pythonjsonlogger            NA\n",
       "pytz                        2025.2\n",
       "referencing                 NA\n",
       "requests                    2.32.4\n",
       "rfc3339_validator           0.1.4\n",
       "rfc3986_validator           0.1.1\n",
       "rpds                        NA\n",
       "scipy                       1.15.3\n",
       "send2trash                  NA\n",
       "six                         1.17.0\n",
       "sklearn                     1.6.1\n",
       "sniffio                     1.3.1\n",
       "stack_data                  0.6.3\n",
       "threadpoolctl               3.6.0\n",
       "tornado                     6.5.1\n",
       "traitlets                   5.14.3\n",
       "typing_extensions           NA\n",
       "typing_inspection           NA\n",
       "uri_template                NA\n",
       "urllib3                     2.4.0\n",
       "wcwidth                     0.2.13\n",
       "webcolors                   NA\n",
       "websocket                   1.8.0\n",
       "yaml                        6.0.2\n",
       "zmq                         26.4.0\n",
       "zoneinfo                    NA\n",
       "</pre>\n",
       "</details> <!-- seems like this ends pre, so might as well be explicit -->\n",
       "<pre>\n",
       "-----\n",
       "IPython             9.2.0\n",
       "jupyter_client      8.6.3\n",
       "jupyter_core        5.7.2\n",
       "jupyterlab          4.4.2\n",
       "-----\n",
       "Python 3.12.10 | packaged by conda-forge | (main, Apr 10 2025, 22:19:24) [Clang 18.1.8 ]\n",
       "macOS-15.3.2-arm64-arm-64bit\n",
       "-----\n",
       "Session information updated at 2025-06-13 09:06\n",
       "</pre>\n",
       "</details>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Session Info\n",
    "session_info.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4d19311e-9227-4f7c-81fd-1dfbc85fd9b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Importing AnnData file\n",
    "# File Path\n",
    "#file_path_Simone = os.path.expanduser(\"~/Göran_Karlsson_Lab/benchLLM/Simone_et_al/7d821d98-5b42-4480-8173-641c1b37b237.h5ad\")\n",
    "#file_path_Granja = os.path.expanduser(\"~/Göran_Karlsson_Lab/benchLLM/Granja_et_al/scRNA-Hematopoiesis-Granja-2019.h5ad\")\n",
    "#file_path_Jardine = os.path.expanduser(\"~/Göran_Karlsson_Lab/benchLLM/Jardine_et_al/fig1b_fbm_scaled_gex_updated_dr_20210104.h5ad\")\n",
    "#file_path_Hiemlich = os.path.expanduser(\"~/Göran_Karlsson_Lab/benchLLM/Heimlich_et_al/5e3e4027-0675-4d5f-b223-bb03a17ade71.h5ad\")\n",
    "#file_path_Roy = os.path.expanduser(\"~/Göran_Karlsson_Lab/benchLLM/Roy_et_al/GSE155259_Roy_et_al.h5ad\")\n",
    "file_path_Zhang = os.path.expanduser(\"~/Göran_Karlsson_Lab/benchLLM/Zhang_et_al/adata_combined_rna_adt_annotated-titration.h5ad\")\n",
    "\n",
    "# AnnData Read In\n",
    "#adata = sc.read_h5ad(file_path_Simone)\n",
    "#adata = sc.read_h5ad(file_path_Hiemlich)\n",
    "#adata = sc.read_h5ad(file_path_Roy)\n",
    "adata = sc.read_h5ad(file_path_Zhang)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "47305bc2-da9f-4fcd-84a2-6ad9a5f2f65a",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'NoneType' object has no attribute 'X'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mAttributeError\u001b[39m                            Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[4]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m adata.X = \u001b[43madata\u001b[49m\u001b[43m.\u001b[49m\u001b[43mraw\u001b[49m\u001b[43m.\u001b[49m\u001b[43mX\u001b[49m.copy()\n",
      "\u001b[31mAttributeError\u001b[39m: 'NoneType' object has no attribute 'X'"
     ]
    }
   ],
   "source": [
    "adata.X = adata.raw.X.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fe284f8c-39a5-498a-a7db-706129a62824",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalize (10,000 Counts per Cell)\n",
    "sc.pp.normalize_total(adata, target_sum=1e4, inplace=True)\n",
    "\n",
    "# Log1p transform\n",
    "sc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "568c4832-f22b-4d98-bdcf-634e3bc62d53",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Min: 0.0\n",
      "Any NaNs in sparse data? False\n",
      "Any Infs in sparse data? False\n"
     ]
    }
   ],
   "source": [
    "# Explore log1p Error (It is due to NaNs)\n",
    "print(\"Min:\", np.min(adata.X))\n",
    "print(\"Any NaNs in sparse data?\", np.isnan(adata.X.data).any())\n",
    "print(\"Any Infs in sparse data?\", np.isinf(adata.X.data).any())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a20a1fe2-837c-4734-805a-98576a4cc668",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:456: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:471: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:482: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:456: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:471: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:482: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:456: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:471: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:482: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:456: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:471: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:482: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:456: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:471: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:482: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:456: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:471: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:482: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:456: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:471: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:482: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:456: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:471: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:482: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:456: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:471: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:482: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:456: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:471: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:482: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:456: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:471: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:482: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:456: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:471: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:482: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n"
     ]
    }
   ],
   "source": [
    "# Zhang\n",
    "# Find Most Up/Down Regulated Genes \n",
    "sc.tl.rank_genes_groups(adata, groupby='Level 2 RNA', method='t-test', key_added = 'Rank_Genes_cell_type')  # try 'wilcoxon' if accurate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "55d052cb-b53a-4d33-82f0-329f3c000c9a",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:456: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:471: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:482: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n"
     ]
    }
   ],
   "source": [
    "# Find Most Up/Down Regulated Genes \n",
    "sc.tl.rank_genes_groups(adata, groupby='cell_type', method='t-test', key_added = 'Rank_Genes_cell_type')  # try 'wilcoxon' if accurate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "535c8c2c-7ebb-4f06-b5d3-137b96116662",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:479: RuntimeWarning: overflow encountered in expm1\n",
      "  foldchanges = (self.expm1_func(mean_group) + 1e-9) / (\n"
     ]
    }
   ],
   "source": [
    "# Jardine\n",
    "# Find Most Up/Down Regulated Genes \n",
    "sc.tl.rank_genes_groups(adata, groupby='broad_fig1_cell.labels', method='t-test', key_added = 'Rank_Genes_cell_type', inplace = True)  # try 'wilcoxon' if accurate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "36a90343-9052-453a-9584-1889f048e19d",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:456: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:471: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:482: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:456: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:471: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:482: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:456: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:471: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:482: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:456: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:471: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:482: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:456: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:471: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:482: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:456: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"names\"] = self.var_names[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:458: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"scores\"] = scores[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:461: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals\"] = pvals[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:471: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"pvals_adj\"] = pvals_adj[global_indices]\n",
      "/Users/alexantill/miniconda3/envs/benchLLM/lib/python3.12/site-packages/scanpy/tools/_rank_genes_groups.py:482: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, \"logfoldchanges\"] = np.log2(\n"
     ]
    }
   ],
   "source": [
    "# Find Most Up/Down Regulated Genes \n",
    "sc.tl.rank_genes_groups(adata, groupby='BioClassification', method='t-test', key_added = 'Rank_Genes_cell_type')  # try 'wilcoxon' if accurate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2d1a5f7e-b80b-45b9-b9cf-2496d9845d07",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"Cell group key 'cell_type' not found in `adata.obs`.\"",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyError\u001b[39m                                  Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[7]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m annotator = \u001b[43mcytetype\u001b[49m\u001b[43m.\u001b[49m\u001b[43mCyteType\u001b[49m\u001b[43m(\u001b[49m\u001b[43madata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgroup_key\u001b[49m\u001b[43m \u001b[49m\u001b[43m=\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mcell_type\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrank_key\u001b[49m\u001b[43m=\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mRank_Genes_cell_type\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgene_symbols_column\u001b[49m\u001b[43m \u001b[49m\u001b[43m=\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mgene_name\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/benchLLM/lib/python3.12/site-packages/cytetype/main.py:109\u001b[39m, in \u001b[36mCyteType.__init__\u001b[39m\u001b[34m(self, adata, group_key, rank_key, gene_symbols_column, n_top_genes, results_prefix, aggregate_metadata, min_percentage, pcent_batch_size)\u001b[39m\n\u001b[32m    106\u001b[39m \u001b[38;5;28mself\u001b[39m.results_prefix = results_prefix\n\u001b[32m    107\u001b[39m \u001b[38;5;28mself\u001b[39m.pcent_batch_size = pcent_batch_size\n\u001b[32m--> \u001b[39m\u001b[32m109\u001b[39m \u001b[43m_validate_adata\u001b[49m\u001b[43m(\u001b[49m\u001b[43madata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgroup_key\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrank_key\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgene_symbols_column\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    111\u001b[39m \u001b[38;5;28mself\u001b[39m.cluster_map = {\n\u001b[32m    112\u001b[39m     \u001b[38;5;28mstr\u001b[39m(x): \u001b[38;5;28mstr\u001b[39m(n + \u001b[32m1\u001b[39m)\n\u001b[32m    113\u001b[39m     \u001b[38;5;28;01mfor\u001b[39;00m n, x \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(\u001b[38;5;28msorted\u001b[39m(adata.obs[group_key].unique().tolist()))\n\u001b[32m    114\u001b[39m }\n\u001b[32m    115\u001b[39m \u001b[38;5;28mself\u001b[39m.clusters = [\n\u001b[32m    116\u001b[39m     \u001b[38;5;28mself\u001b[39m.cluster_map[\u001b[38;5;28mstr\u001b[39m(x)] \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m adata.obs[group_key].values.tolist()\n\u001b[32m    117\u001b[39m ]\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/benchLLM/lib/python3.12/site-packages/cytetype/anndata_helpers.py:17\u001b[39m, in \u001b[36m_validate_adata\u001b[39m\u001b[34m(adata, cell_group_key, rank_genes_key, gene_symbols_col)\u001b[39m\n\u001b[32m     14\u001b[39m \u001b[38;5;250m\u001b[39m\u001b[33;03m\"\"\"Validate the AnnData object structure.\"\"\"\u001b[39;00m\n\u001b[32m     16\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m cell_group_key \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m adata.obs:\n\u001b[32m---> \u001b[39m\u001b[32m17\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mCell group key \u001b[39m\u001b[33m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mcell_group_key\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m'\u001b[39m\u001b[33m not found in `adata.obs`.\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m     18\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m adata.X \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m     19\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[32m     20\u001b[39m         \u001b[33m\"\u001b[39m\u001b[33m`adata.X` is required for ranking genes. Please ensure it contains log1p normalized data.\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m     21\u001b[39m     )\n",
      "\u001b[31mKeyError\u001b[39m: \"Cell group key 'cell_type' not found in `adata.obs`.\""
     ]
    }
   ],
   "source": [
    "annotator = cytetype.CyteType(adata, group_key = 'cell_type', rank_key= 'Rank_Genes_cell_type', gene_symbols_column = 'gene_name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "03ddee14-26c2-4a74-af3d-8f094603d44f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating expression percentages.\n",
      "Extracting marker genes.\n",
      "Data preparation completed. Ready for submitting jobs.\n"
     ]
    }
   ],
   "source": [
    "adata.var[\"gene_symbols\"] = adata.var.index\n",
    "annotator = cytetype.CyteType(adata, group_key = 'cell_type', rank_key= 'Rank_Genes_cell_type', gene_symbols_column = 'gene_symbols')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "734cb2c1-2efd-499a-ba6c-2bba5286f2a9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating expression percentages.\n",
      "Extracting marker genes.\n",
      "Data preparation completed. Ready for submitting jobs.\n"
     ]
    }
   ],
   "source": [
    "# Zhang\n",
    "adata.var[\"gene_symbols\"] = adata.var.index\n",
    "annotator = cytetype.CyteType(adata, group_key = 'Level 2 RNA', rank_key= 'Rank_Genes_cell_type', gene_symbols_column = 'gene_symbols')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "54d7b081-1719-4ac0-a015-401ca770920b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating expression percentages.\n",
      "Extracting marker genes.\n",
      "Data preparation completed. Ready for submitting jobs.\n"
     ]
    }
   ],
   "source": [
    "#Roy\n",
    "annotator = cytetype.CyteType(adata, group_key = 'cell_type', rank_key= 'Rank_Genes_cell_type', gene_symbols_column = 'name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "03ee8358-9b5d-40fe-aadd-2bb6a563fd7e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating expression percentages.\n",
      "Extracting marker genes.\n",
      "Data preparation completed. Ready for submitting jobs.\n"
     ]
    }
   ],
   "source": [
    "adata.var[\"gene_symbols\"] = adata.var.index\n",
    "annotator = cytetype.CyteType(adata, group_key = 'broad_fig1_cell.labels', rank_key= 'Rank_Genes_cell_type', gene_symbols_column = 'gene_symbols')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "cfcf0b1c-7c38-4c0b-ae8e-0c1722aa4a9d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating expression percentages.\n",
      "Extracting marker genes.\n",
      "Data preparation completed. Ready for submitting jobs.\n"
     ]
    }
   ],
   "source": [
    "annotator = cytetype.CyteType(adata, group_key = 'BioClassification', rank_key= 'Rank_Genes_cell_type', gene_symbols_column = 'name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "c651ffee-c15a-4989-8cc5-1b9ad41e76f3",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true,
     "source_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Waiting for results for job ID: 320eac1b-2e03-40c3-844e-336a58ab3db7\n",
      "View the automatically updating visualization report at: https://nygen-labs-prod--cell-annotation-agent-fastapi-app.modal.run/report/320eac1b-2e03-40c3-844e-336a58ab3db7\n",
      "250605:1154:22 |INFO| [STARTING WORKFLOWS]: Job ID: 320eac1b-2e03-40c3-844e-336a58ab3db7\n",
      "250605:1154:22 |INFO| [WORKFLOW STEP 1/3]: Generating context summary\n",
      "250605:1155:41 |INFO| [WORKFLOW STEP 2/3]: Running cluster contextualization, annotation, review, and ontology assignment\n",
      "250605:1155:43 |INFO| [Cluster: 1] Generating cluster-specific context...\n",
      "250605:1155:46 |INFO| [Cluster: 10] Generating cluster-specific context...\n",
      "250605:1155:51 |INFO| [Cluster: 2] Generating cluster-specific context...\n",
      "250605:1155:52 |INFO| [Cluster: 3] Generating cluster-specific context...\n",
      "250605:1155:53 |INFO| [Cluster: 4] Generating cluster-specific context...\n",
      "250605:1156:21 |INFO| [Cluster: 3] Annotating cluster...\n",
      "250605:1156:23 |INFO| [Cluster: 2] Annotating cluster...\n",
      "250605:1156:42 |INFO| [Cluster: 1] Annotating cluster...\n",
      "250605:1156:55 |INFO| [Cluster: 10] Annotating cluster...\n",
      "250605:1157:03 |INFO| [Cluster: 3] Reviewing cluster annotation...\n",
      "250605:1157:05 |INFO| [Cluster: 4] Annotating cluster...\n",
      "250605:1157:34 |INFO| [Cluster: 1] Reviewing cluster annotation...\n",
      "250605:1157:40 |INFO| [Cluster: 3] Finding ontology term for annotation...\n",
      "250605:1158:43 |INFO| [Cluster: 1] Finding ontology term for annotation...\n",
      "250605:1158:45 |INFO| [Cluster: 2] Reviewing cluster annotation...\n",
      "250605:1158:59 |SUCCESS| [Cluster: 3] Annotation (qwen/qwen3-235b-a22b) completed in 186.97s\n",
      "250605:1158:59 |INFO| [Cluster: 3] Saving partial result (1/10)\n",
      "250605:1159:02 |INFO| [Cluster: 5] Generating cluster-specific context...\n",
      "250605:1159:14 |INFO| [Cluster: 2] ReAnnotating using reviewer feedback\n",
      "250605:1159:23 |INFO| [Cluster: 5] Annotating cluster...\n",
      "250605:1159:25 |SUCCESS| [Cluster: 1] Annotation (qwen/qwen3-235b-a22b) completed in 222.32s\n",
      "250605:1159:25 |INFO| [Cluster: 1] Saving partial result (2/10)\n",
      "250605:1159:28 |INFO| [Cluster: 6] Generating cluster-specific context...\n",
      "250605:1200:07 |INFO| [Cluster: 5] Reviewing cluster annotation...\n",
      "250605:1200:13 |INFO| [Cluster: 4] Reviewing cluster annotation...\n",
      "250605:1200:48 |INFO| [Cluster: 5] Finding ontology term for annotation...\n",
      "250605:1200:54 |INFO| [Cluster: 10] Reviewing cluster annotation...\n",
      "250605:1200:57 |INFO| [Cluster: 6] Annotating cluster...\n",
      "250605:1201:22 |INFO| [Cluster: 2] Reviewing cluster annotation...\n",
      "250605:1201:33 |SUCCESS| [Cluster: 5] Annotation (qwen/qwen3-235b-a22b) completed in 150.70s\n",
      "250605:1201:33 |INFO| [Cluster: 5] Saving partial result (3/10)\n",
      "250605:1201:35 |INFO| [Cluster: 7] Generating cluster-specific context...\n",
      "250605:1201:35 |INFO| [Cluster: 4] Finding ontology term for annotation...\n",
      "250605:1201:51 |INFO| [Cluster: 10] Finding ontology term for annotation...\n",
      "250605:1202:06 |INFO| [Cluster: 7] Annotating cluster...\n",
      "250605:1202:26 |SUCCESS| [Cluster: 4] Annotation (qwen/qwen3-235b-a22b) completed in 393.37s\n",
      "250605:1202:26 |INFO| [Cluster: 4] Saving partial result (4/10)\n",
      "250605:1202:28 |INFO| [Cluster: 8] Generating cluster-specific context...\n",
      "250605:1202:58 |INFO| [Cluster: 2] Finding ontology term for annotation...\n",
      "250605:1203:05 |INFO| [Cluster: 8] Annotating cluster...\n",
      "250605:1203:15 |SUCCESS| [Cluster: 10] Annotation (qwen/qwen3-235b-a22b) completed in 448.07s\n",
      "250605:1203:15 |INFO| [Cluster: 10] Saving partial result (5/10)\n",
      "250605:1203:16 |INFO| [Cluster: 9] Generating cluster-specific context...\n",
      "250605:1203:42 |SUCCESS| [Cluster: 2] Annotation (qwen/qwen3-235b-a22b) completed in 470.41s\n",
      "250605:1203:42 |INFO| [Cluster: 2] Saving partial result (6/10)\n",
      "250605:1203:59 |INFO| [Cluster: 9] Annotating cluster...\n",
      "250605:1204:12 |INFO| [Cluster: 8] Reviewing cluster annotation...\n",
      "250605:1204:34 |INFO| [Cluster: 6] Reviewing cluster annotation...\n",
      "250605:1205:15 |INFO| [Cluster: 7] Reviewing cluster annotation...\n",
      "250605:1205:34 |INFO| [Cluster: 6] Finding ontology term for annotation...\n",
      "250605:1206:06 |INFO| [Cluster: 8] Finding ontology term for annotation...\n",
      "250605:1206:23 |INFO| [Cluster: 7] Finding ontology term for annotation...\n",
      "250605:1206:30 |SUCCESS| [Cluster: 6] Annotation (qwen/qwen3-235b-a22b) completed in 421.70s\n",
      "250605:1206:30 |INFO| [Cluster: 6] Saving partial result (7/10)\n",
      "250605:1207:29 |INFO| [Cluster: 9] Reviewing cluster annotation...\n",
      "250605:1208:06 |SUCCESS| [Cluster: 7] Annotation (qwen/qwen3-235b-a22b) completed in 390.70s\n",
      "250605:1208:06 |INFO| [Cluster: 7] Saving partial result (8/10)\n",
      "250605:1210:10 |INFO| [Cluster: 9] ReAnnotating using reviewer feedback\n",
      "250605:1210:33 |SUCCESS| [Cluster: 8] Annotation (qwen/qwen3-235b-a22b) completed in 485.56s\n",
      "250605:1210:33 |INFO| [Cluster: 8] Saving partial result (9/10)\n",
      "250605:1213:58 |INFO| [Cluster: 9] Reviewing cluster annotation...\n",
      "250605:1214:38 |INFO| [Cluster: 9] Finding ontology term for annotation...\n",
      "250605:1216:16 |SUCCESS| [Cluster: 9] Annotation (qwen/qwen3-235b-a22b) completed in 779.85s\n",
      "250605:1216:16 |INFO| [Cluster: 9] Saving partial result (10/10)\n",
      "250605:1216:17 |INFO| [WORKFLOW STEP 3/3]: Generating summary\n",
      "Job 320eac1b-2e03-40c3-844e-336a58ab3db7 completed successfully.\n",
      "Annotations successfully added to `adata.obs['cytetype_broad_fig1_cell.labels']` and `adata.uns['cytetype_results']`.\n"
     ]
    }
   ],
   "source": [
    "# Run Annotation with custom model\n",
    "adata = annotator.run(\n",
    "    study_context = \"Healthy human fetal bone marrow cell, analysed 10x genomics single cell RNA-seq kits\",\n",
    "    model_config = [{\n",
    "        'provider': 'openai',\n",
    "        'name': 'qwen/qwen3-235b-a22b',\n",
    "        'apiKey': os.environ.get('OPENROUTER_API_KEY'),\n",
    "        'baseUrl': \"https://openrouter.ai/api/v1\",\n",
    "    }],\n",
    "    auth_token = os.environ.get('NYGEN_API_KEY'),\n",
    "    timeout_seconds=3000,\n",
    "    run_config={\n",
    "        'concurrentClusters': 5,        # Default: 3, Range: 2-10\n",
    "        'maxAnnotationRevisions': 2,    # Default: 2, Range: 1-5\n",
    "        'maxLLMRequests': 250           # Default: 500, Range: 50-2000\n",
    "    },\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0afa8935-f45c-4dd3-a8d4-79190f440bfd",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true,
     "source_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Waiting for results for job ID: 54d0c1f1-d922-4c24-bae0-e43930fbee49\n",
      "View the automatically updating visualization report at: https://nygen-labs-prod--cell-annotation-agent-fastapi-app.modal.run/report/54d0c1f1-d922-4c24-bae0-e43930fbee49\n",
      "Could not fetch logs for job 54d0c1f1-d922-4c24-bae0-e43930fbee49: 404 Client Error: Not Found for url: https://nygen-labs-prod--cell-annotation-agent-fastapi-app.modal.run/logs/54d0c1f1-d922-4c24-bae0-e43930fbee49\n",
      "250605:1003:46 |INFO| [STARTING WORKFLOWS]: Job ID: 54d0c1f1-d922-4c24-bae0-e43930fbee49\n",
      "250605:1003:46 |INFO| [WORKFLOW STEP 1/3]: Generating context summary\n",
      "250605:1004:42 |INFO| [WORKFLOW STEP 2/3]: Running cluster contextualization, annotation, review, and ontology assignment\n",
      "250605:1004:46 |INFO| [Cluster: 1] Generating cluster-specific context...\n",
      "250605:1004:50 |INFO| [Cluster: 10] Generating cluster-specific context...\n",
      "250605:1004:55 |INFO| [Cluster: 2] Generating cluster-specific context...\n",
      "250605:1004:59 |INFO| [Cluster: 3] Generating cluster-specific context...\n",
      "250605:1005:03 |INFO| [Cluster: 4] Generating cluster-specific context...\n",
      "250605:1005:33 |INFO| [Cluster: 4] Annotating cluster...\n",
      "250605:1005:46 |INFO| [Cluster: 10] Annotating cluster...\n",
      "250605:1006:36 |INFO| [Cluster: 3] Annotating cluster...\n",
      "250605:1006:38 |INFO| [Cluster: 2] Annotating cluster...\n",
      "250605:1006:39 |INFO| [Cluster: 1] Annotating cluster...\n",
      "250605:1008:06 |INFO| [Cluster: 4] Reviewing cluster annotation...\n",
      "250605:1009:01 |INFO| [Cluster: 4] Finding ontology term for annotation...\n",
      "250605:1009:21 |INFO| [Cluster: 1] Reviewing cluster annotation...\n",
      "250605:1009:34 |INFO| [Cluster: 2] Reviewing cluster annotation...\n",
      "250605:1009:57 |INFO| [Cluster: 1] Finding ontology term for annotation...\n",
      "250605:1010:02 |SUCCESS| [Cluster: 4] Annotation (qwen/qwen3-235b-a22b) completed in 298.62s\n",
      "250605:1010:02 |INFO| [Cluster: 4] Saving partial result (1/10)\n",
      "250605:1010:12 |INFO| [Cluster: 5] Generating cluster-specific context...\n",
      "250605:1010:12 |INFO| [Cluster: 10] Reviewing cluster annotation...\n",
      "250605:1010:22 |INFO| [Cluster: 2] Finding ontology term for annotation...\n",
      "250605:1010:53 |SUCCESS| [Cluster: 2] Annotation (qwen/qwen3-235b-a22b) completed in 358.55s\n",
      "250605:1010:53 |INFO| [Cluster: 2] Saving partial result (2/10)\n",
      "250605:1011:03 |INFO| [Cluster: 6] Generating cluster-specific context...\n",
      "250605:1011:03 |INFO| [Cluster: 10] Finding ontology term for annotation...\n",
      "250605:1011:32 |INFO| [Cluster: 5] Annotating cluster...\n",
      "250605:1011:34 |INFO| [Cluster: 3] Reviewing cluster annotation...\n",
      "250605:1011:37 |SUCCESS| [Cluster: 1] Annotation () completed in 411.46s\n",
      "250605:1011:37 |INFO| [Cluster: 1] Saving partial result (3/10)\n",
      "250605:1011:47 |INFO| [Cluster: 7] Generating cluster-specific context...\n",
      "250605:1011:55 |SUCCESS| [Cluster: 10] Annotation (qwen/qwen3-235b-a22b) completed in 424.06s\n",
      "250605:1011:55 |INFO| [Cluster: 10] Saving partial result (4/10)\n",
      "250605:1012:04 |INFO| [Cluster: 8] Generating cluster-specific context...\n",
      "250605:1012:21 |INFO| [Cluster: 7] Annotating cluster...\n",
      "250605:1012:29 |INFO| [Cluster: 3] Finding ontology term for annotation...\n",
      "250605:1012:38 |INFO| [Cluster: 6] Annotating cluster...\n",
      "250605:1012:59 |INFO| [Cluster: 8] Annotating cluster...\n",
      "250605:1013:36 |SUCCESS| [Cluster: 3] Annotation (qwen/qwen3-235b-a22b) completed in 516.73s\n",
      "250605:1013:36 |INFO| [Cluster: 3] Saving partial result (5/10)\n",
      "250605:1013:45 |INFO| [Cluster: 9] Generating cluster-specific context...\n",
      "250605:1014:05 |INFO| [Cluster: 5] Reviewing cluster annotation...\n",
      "250605:1014:31 |INFO| [Cluster: 9] Annotating cluster...\n",
      "250605:1014:32 |INFO| [Cluster: 7] Reviewing cluster annotation...\n",
      "250605:1014:58 |INFO| [Cluster: 6] Reviewing cluster annotation...\n",
      "250605:1015:23 |INFO| [Cluster: 7] Finding ontology term for annotation...\n",
      "250605:1015:23 |INFO| [Cluster: 5] Finding ontology term for annotation...\n",
      "250605:1016:05 |INFO| [Cluster: 6] ReAnnotating using reviewer feedback\n",
      "250605:1016:06 |INFO| [Cluster: 9] Reviewing cluster annotation...\n",
      "250605:1016:27 |SUCCESS| [Cluster: 7] Annotation (qwen/qwen3-235b-a22b) completed in 280.36s\n",
      "250605:1016:27 |INFO| [Cluster: 7] Saving partial result (6/10)\n",
      "250605:1016:35 |SUCCESS| [Cluster: 5] Annotation (qwen/qwen3-235b-a22b) completed in 383.32s\n",
      "250605:1016:35 |INFO| [Cluster: 5] Saving partial result (7/10)\n",
      "250605:1016:58 |INFO| [Cluster: 8] Reviewing cluster annotation...\n",
      "250605:1017:47 |INFO| [Cluster: 8] Finding ontology term for annotation...\n",
      "250605:1017:56 |INFO| [Cluster: 9] Finding ontology term for annotation...\n",
      "250605:1018:07 |INFO| [Cluster: 6] Reviewing cluster annotation...\n",
      "250605:1018:33 |SUCCESS| [Cluster: 9] Annotation (qwen/qwen3-235b-a22b) completed in 287.95s\n",
      "250605:1018:33 |INFO| [Cluster: 9] Saving partial result (8/10)\n",
      "250605:1019:38 |SUCCESS| [Cluster: 8] Annotation (qwen/qwen3-235b-a22b) completed in 453.84s\n",
      "250605:1019:38 |INFO| [Cluster: 8] Saving partial result (9/10)\n",
      "250605:1019:55 |INFO| [Cluster: 6] Finding ontology term for annotation...\n",
      "250605:1020:52 |SUCCESS| [Cluster: 6] Annotation (qwen/qwen3-235b-a22b) completed in 589.05s\n",
      "250605:1020:52 |INFO| [Cluster: 6] Saving partial result (10/10)\n",
      "250605:1020:57 |INFO| [WORKFLOW STEP 3/3]: Generating summary\n",
      "Job 54d0c1f1-d922-4c24-bae0-e43930fbee49 completed successfully.\n",
      "Annotations successfully added to `adata.obs['cytetype_broad_fig1_cell.labels']` and `adata.uns['cytetype_results']`.\n"
     ]
    }
   ],
   "source": [
    "# Run annotation with custom model\n",
    "adata = annotator.run(\n",
    "    study_context = \"Human adult mixed-phenotype acute leukemias from peripheral blood and bone marrow, analysed 10x genomics single cell RNA-seq kits\",\n",
    "    model_config = [{\n",
    "        'provider': 'openai',\n",
    "        'name': 'qwen/qwen3-235b-a22b',\n",
    "        'apiKey': os.environ.get('OPENROUTER_API_KEY'),\n",
    "        'baseUrl': \"https://openrouter.ai/api/v1\",\n",
    "    }],\n",
    "    auth_token = os.environ.get('NYGEN_API_KEY'),\n",
    "    timeout_seconds=3000,\n",
    "    run_config={\n",
    "        'concurrentClusters': 5,        # Default: 3, Range: 2-10\n",
    "        'maxAnnotationRevisions': 2,    # Default: 2, Range: 1-5\n",
    "        'maxLLMRequests': 250           # Default: 500, Range: 50-2000\n",
    "    },\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "506637d5-77ec-4b4e-aa11-f1fff49a0dd8",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true,
     "source_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Waiting for results for job ID: 6248f067-9da2-444c-8ee5-526186ab2ef0\n",
      "View the automatically updating visualization report at: https://nygen-labs-prod--cell-annotation-agent-fastapi-app.modal.run/report/6248f067-9da2-444c-8ee5-526186ab2ef0\n",
      "250610:1046:11 |INFO| [STARTING WORKFLOWS]: Job ID: 6248f067-9da2-444c-8ee5-526186ab2ef0\n",
      "250610:1046:11 |INFO| [WORKFLOW STEP 1/3]: Generating context summary\n",
      "250610:1046:18 |INFO| [WORKFLOW STEP 2/3]: Running complete annotation pipeline for each cluster\n",
      "250610:1046:19 |INFO| [Cluster: 1] Generating cluster-specific context...\n",
      "250610:1046:19 |INFO| [Cluster: 2] Generating cluster-specific context...\n",
      "250610:1046:20 |INFO| [Cluster: 3] Generating cluster-specific context...\n",
      "250610:1046:21 |INFO| [Cluster: 4] Generating cluster-specific context...\n",
      "250610:1046:22 |INFO| [Cluster: 5] Generating cluster-specific context...\n",
      "250610:1046:24 |INFO| [Cluster: 1] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1046:24 |INFO| [Cluster: 4] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1046:24 |INFO| [Cluster: 5] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1046:24 |INFO| [Cluster: 3] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1046:24 |INFO| [Cluster: 2] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1046:30 |INFO| [Cluster: 4] Finding ontology term for annotation...\n",
      "250610:1046:32 |INFO| [Cluster: 5] Finding ontology term for annotation...\n",
      "250610:1046:33 |INFO| [Cluster: 4] Fetching CellGuide markers...\n",
      "250610:1046:34 |INFO| [Cluster: 5] Fetching CellGuide markers...\n",
      "250610:1046:35 |INFO| [Cluster: 4] Final review with CellGuide markers...\n",
      "250610:1046:36 |INFO| [Cluster: 1] Finding ontology term for annotation...\n",
      "250610:1046:36 |INFO| [Cluster: 4] Annotation approved by reviewer\n",
      "250610:1046:36 |SUCCESS| [Cluster: 4] Annotation pipeline completed in 15.43s after 1 attempt(s)\n",
      "250610:1046:36 |INFO| [Cluster: 4] LLM Usage: 8 requests, 25152 input tokens, 1258 output tokens, 26410 total tokens\n",
      "250610:1046:36 |INFO| [Cluster: 4] Saving partial result (1/9)\n",
      "250610:1046:38 |INFO| [Cluster: 6] Generating cluster-specific context...\n",
      "250610:1046:38 |INFO| [Cluster: 5] Final review with CellGuide markers...\n",
      "250610:1046:38 |INFO| [Cluster: 2] Finding ontology term for annotation...\n",
      "250610:1046:40 |INFO| [Cluster: 6] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1046:40 |INFO| [Cluster: 5] Annotation not approved, will retry with feedback: ...\n",
      "250610:1046:40 |INFO| [Cluster: 5] Attempt 2: Running complete annotation pipeline...\n",
      "250610:1046:42 |INFO| [Cluster: 1] Fetching CellGuide markers...\n",
      "250610:1046:44 |INFO| [Cluster: 1] Final review with CellGuide markers...\n",
      "250610:1046:46 |INFO| [Cluster: 2] Fetching CellGuide markers...\n",
      "250610:1046:48 |INFO| [Cluster: 2] Final review with CellGuide markers...\n",
      "250610:1046:53 |INFO| [Cluster: 2] Annotation approved by reviewer\n",
      "250610:1046:53 |SUCCESS| [Cluster: 2] Annotation pipeline completed in 33.06s after 1 attempt(s)\n",
      "250610:1046:53 |INFO| [Cluster: 2] LLM Usage: 15 requests, 50989 input tokens, 2137 output tokens, 53126 total tokens\n",
      "250610:1046:53 |INFO| [Cluster: 2] Saving partial result (2/9)\n",
      "250610:1046:54 |INFO| [Cluster: 7] Generating cluster-specific context...\n",
      "250610:1046:54 |INFO| [Cluster: 5] Finding ontology term for annotation...\n",
      "250610:1046:56 |INFO| [Cluster: 7] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1046:58 |INFO| [Cluster: 6] Finding ontology term for annotation...\n",
      "250610:1046:59 |INFO| [Cluster: 5] Fetching CellGuide markers...\n",
      "250610:1047:03 |INFO| [Cluster: 6] Fetching CellGuide markers...\n",
      "250610:1047:09 |INFO| [Cluster: 1] Annotation approved by reviewer\n",
      "250610:1047:09 |SUCCESS| [Cluster: 1] Annotation pipeline completed in 49.79s after 1 attempt(s)\n",
      "250610:1047:09 |INFO| [Cluster: 1] LLM Usage: 15 requests, 58181 input tokens, 3004 output tokens, 61185 total tokens\n",
      "250610:1047:09 |INFO| [Cluster: 1] Saving partial result (3/9)\n",
      "250610:1047:10 |INFO| [Cluster: 8] Generating cluster-specific context...\n",
      "250610:1047:12 |INFO| [Cluster: 7] Finding ontology term for annotation...\n",
      "250610:1047:12 |INFO| [Cluster: 8] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1047:13 |INFO| [Cluster: 5] Final review with CellGuide markers...\n",
      "250610:1047:14 |INFO| [Cluster: 5] Reached max retries (2), proceeding with current annotation\n",
      "250610:1047:14 |SUCCESS| [Cluster: 5] Annotation pipeline completed in 52.22s after 2 attempt(s)\n",
      "250610:1047:14 |INFO| [Cluster: 5] LLM Usage: 14 requests, 55863 input tokens, 3985 output tokens, 59848 total tokens\n",
      "250610:1047:14 |INFO| [Cluster: 5] Saving partial result (4/9)\n",
      "250610:1047:16 |INFO| [Cluster: 9] Generating cluster-specific context...\n",
      "250610:1047:18 |INFO| [Cluster: 6] Final review with CellGuide markers...\n",
      "250610:1047:19 |INFO| [Cluster: 9] Attempt 1: Running complete annotation pipeline...\n",
      "Job 6248f067-9da2-444c-8ee5-526186ab2ef0 completed successfully.\n",
      "Annotations successfully added to `adata.obs['cytetype_annotation_cell_type']` , ontology term added to `adata.obs['cytetype_cellOntologyTerm_cell_type']` and, full results added to `adata.uns['cytetype_results']`.\n"
     ]
    }
   ],
   "source": [
    "# Heimlich\n",
    "# Run annotation with custom model\n",
    "adata = annotator.run(\n",
    "    study_context = \"Human adult with clonal hematopoiesis or healthy peripheral blood, analysed 10x genomics single cell RNA-seq kits\",\n",
    "    model_config = [{\n",
    "        'provider': 'openai',\n",
    "        'name': 'meta-llama/llama-4-maverick',\n",
    "        'apiKey': os.environ.get('OPENROUTER_API_KEY'),\n",
    "        'baseUrl': \"https://openrouter.ai/api/v1\",\n",
    "    }],\n",
    "    auth_token = os.environ.get('NYGEN_API_KEY'),\n",
    "    timeout_seconds=3000,\n",
    "    run_config={\n",
    "        'concurrentClusters': 5,        # Default: 3, Range: 2-10\n",
    "        'maxAnnotationRevisions': 2,    # Default: 2, Range: 1-5\n",
    "        'maxLLMRequests': 225           # Default: 500, Range: 50-2000\n",
    "    },\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "34437a2f-64e1-4feb-af2f-1daf593184d9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Waiting for results for job ID: ba82f2af-cf37-4548-ac39-be56cc2441b5\n",
      "View the automatically updating visualization report at: https://nygen-labs-prod--cell-annotation-agent-fastapi-app.modal.run/report/ba82f2af-cf37-4548-ac39-be56cc2441b5\n",
      "250613:0712:02 |INFO| [STARTING WORKFLOWS]: Job ID: ba82f2af-cf37-4548-ac39-be56cc2441b5\n",
      "250613:0712:02 |INFO| [WORKFLOW STEP 1/3]: Generating context summary\n",
      "250613:0713:31 |INFO| [WORKFLOW STEP 2/3]: Running complete annotation pipeline for each cluster\n",
      "250613:0713:32 |INFO| [Cluster: 1] Generating cluster-specific context...\n",
      "250613:0713:33 |INFO| [Cluster: 10] Generating cluster-specific context...\n",
      "250613:0713:35 |INFO| [Cluster: 11] Generating cluster-specific context...\n",
      "250613:0713:36 |INFO| [Cluster: 12] Generating cluster-specific context...\n",
      "250613:0713:36 |INFO| [Cluster: 13] Generating cluster-specific context...\n",
      "250613:0714:27 |INFO| [Cluster: 10] Attempt 1: Running complete annotation pipeline...\n",
      "250613:0714:30 |INFO| [Cluster: 13] Attempt 1: Running complete annotation pipeline...\n",
      "250613:0714:35 |INFO| [Cluster: 1] Attempt 1: Running complete annotation pipeline...\n",
      "250613:0714:37 |INFO| [Cluster: 11] Attempt 1: Running complete annotation pipeline...\n",
      "250613:0714:51 |INFO| [Cluster: 12] Attempt 1: Running complete annotation pipeline...\n",
      "250613:0715:38 |INFO| [Cluster: 1] Finding ontology term for annotation...\n",
      "250613:0715:39 |INFO| [Cluster: 13] Finding ontology term for annotation...\n",
      "250613:0715:41 |INFO| [Cluster: 10] Finding ontology term for annotation...\n",
      "250613:0715:43 |INFO| [Cluster: 11] Finding ontology term for annotation...\n",
      "250613:0715:50 |INFO| [Cluster: 12] Finding ontology term for annotation...\n",
      "250613:0716:36 |INFO| [Cluster: 1] Fetching CellGuide markers...\n",
      "250613:0716:36 |INFO| [Cluster: 1] Final review with CellGuide markers...\n",
      "250613:0716:41 |INFO| [Cluster: 11] Fetching CellGuide markers...\n",
      "250613:0716:41 |INFO| [Cluster: 11] Final review with CellGuide markers...\n",
      "250613:0716:46 |INFO| [Cluster: 10] Fetching CellGuide markers...\n",
      "250613:0716:46 |INFO| [Cluster: 10] Final review with CellGuide markers...\n",
      "250613:0716:50 |INFO| [Cluster: 13] Fetching CellGuide markers...\n",
      "250613:0716:50 |INFO| [Cluster: 13] Final review with CellGuide markers...\n",
      "250613:0716:56 |INFO| [Cluster: 12] Fetching CellGuide markers...\n",
      "250613:0716:56 |INFO| [Cluster: 12] Final review with CellGuide markers...\n",
      "250613:0717:32 |INFO| [Cluster: 1] Annotation not approved, will retry with feedback: Reviewer agent failed due to error: status_code: 403, model_name: qwen/qwen3-235b-a22b, body: {'mess...\n",
      "250613:0717:32 |INFO| [Cluster: 1] Retry 2: Re-running complete pipeline with feedback\n",
      "250613:0717:41 |INFO| [Cluster: 13] Annotation not approved, will retry with feedback: Reviewer agent failed due to error: status_code: 403, model_name: qwen/qwen3-235b-a22b, body: {'mess...\n",
      "250613:0717:41 |INFO| [Cluster: 13] Retry 2: Re-running complete pipeline with feedback\n",
      "250613:0717:44 |INFO| [Cluster: 11] Annotation not approved, will retry with feedback: Reviewer agent failed due to error: status_code: 403, model_name: qwen/qwen3-235b-a22b, body: {'mess...\n",
      "250613:0717:44 |INFO| [Cluster: 11] Retry 2: Re-running complete pipeline with feedback\n",
      "250613:0717:52 |INFO| [Cluster: 12] Annotation not approved, will retry with feedback: Reviewer agent failed due to error: status_code: 403, model_name: qwen/qwen3-235b-a22b, body: {'mess...\n",
      "250613:0717:52 |INFO| [Cluster: 12] Retry 2: Re-running complete pipeline with feedback\n",
      "250613:0717:57 |INFO| [Cluster: 10] Annotation not approved, will retry with feedback: Reviewer agent failed due to error: status_code: 403, model_name: qwen/qwen3-235b-a22b, body: {'mess...\n",
      "250613:0717:57 |INFO| [Cluster: 10] Retry 2: Re-running complete pipeline with feedback\n",
      "250613:0718:38 |INFO| [Cluster: 11] Finding ontology term for annotation...\n",
      "250613:0718:42 |INFO| [Cluster: 1] Finding ontology term for annotation...\n",
      "250613:0718:44 |INFO| [Cluster: 10] Finding ontology term for annotation...\n",
      "250613:0718:56 |INFO| [Cluster: 12] Finding ontology term for annotation...\n",
      "250613:0718:57 |INFO| [Cluster: 13] Finding ontology term for annotation...\n",
      "250613:0719:42 |INFO| [Cluster: 10] Fetching CellGuide markers...\n",
      "250613:0719:42 |INFO| [Cluster: 10] Final review with CellGuide markers...\n",
      "250613:0719:45 |INFO| [Cluster: 12] Fetching CellGuide markers...\n",
      "250613:0719:45 |INFO| [Cluster: 12] Final review with CellGuide markers...\n",
      "250613:0719:51 |INFO| [Cluster: 11] Fetching CellGuide markers...\n",
      "250613:0719:51 |INFO| [Cluster: 11] Final review with CellGuide markers...\n",
      "250613:0719:53 |INFO| [Cluster: 1] Fetching CellGuide markers...\n",
      "250613:0719:53 |INFO| [Cluster: 1] Final review with CellGuide markers...\n",
      "250613:0720:05 |INFO| [Cluster: 13] Fetching CellGuide markers...\n",
      "250613:0720:05 |INFO| [Cluster: 13] Final review with CellGuide markers...\n",
      "250613:0720:52 |INFO| [Cluster: 12] Reached max retries (2), proceeding with current annotation\n",
      "250613:0720:52 |SUCCESS| [Cluster: 12] Annotation pipeline completed in 436.29s after 2 attempt(s)\n",
      "250613:0720:52 |INFO| [Cluster: 12] LLM Usage: 0 requests, 0 input tokens, 0 output tokens, 0 total tokens\n",
      "250613:0720:52 |INFO| [Cluster: 12] Saving partial result (1/32)\n",
      "250613:0720:54 |INFO| [Cluster: 14] Generating cluster-specific context...\n",
      "250613:0720:54 |INFO| [Cluster: 11] Reached max retries (2), proceeding with current annotation\n",
      "250613:0720:54 |SUCCESS| [Cluster: 11] Annotation pipeline completed in 439.50s after 2 attempt(s)\n",
      "250613:0720:54 |INFO| [Cluster: 11] LLM Usage: 0 requests, 0 input tokens, 0 output tokens, 0 total tokens\n",
      "250613:0720:54 |INFO| [Cluster: 11] Saving partial result (2/32)\n",
      "250613:0720:56 |INFO| [Cluster: 15] Generating cluster-specific context...\n",
      "250613:0720:57 |INFO| [Cluster: 10] Reached max retries (2), proceeding with current annotation\n",
      "250613:0720:57 |SUCCESS| [Cluster: 10] Annotation pipeline completed in 443.48s after 2 attempt(s)\n",
      "250613:0720:57 |INFO| [Cluster: 10] LLM Usage: 0 requests, 0 input tokens, 0 output tokens, 0 total tokens\n",
      "250613:0720:57 |INFO| [Cluster: 10] Saving partial result (3/32)\n",
      "250613:0720:58 |INFO| [Cluster: 16] Generating cluster-specific context...\n",
      "250613:0720:59 |INFO| [Cluster: 1] Reached max retries (2), proceeding with current annotation\n",
      "250613:0720:59 |SUCCESS| [Cluster: 1] Annotation pipeline completed in 446.41s after 2 attempt(s)\n",
      "250613:0720:59 |INFO| [Cluster: 1] LLM Usage: 0 requests, 0 input tokens, 0 output tokens, 0 total tokens\n",
      "250613:0720:59 |INFO| [Cluster: 1] Saving partial result (4/32)\n",
      "250613:0721:01 |INFO| [Cluster: 17] Generating cluster-specific context...\n",
      "250613:0721:08 |INFO| [Cluster: 13] Reached max retries (2), proceeding with current annotation\n",
      "250613:0721:08 |SUCCESS| [Cluster: 13] Annotation pipeline completed in 451.51s after 2 attempt(s)\n",
      "250613:0721:08 |INFO| [Cluster: 13] LLM Usage: 0 requests, 0 input tokens, 0 output tokens, 0 total tokens\n",
      "250613:0721:08 |INFO| [Cluster: 13] Saving partial result (5/32)\n",
      "250613:0721:09 |INFO| [Cluster: 18] Generating cluster-specific context...\n",
      "250613:0721:54 |INFO| [Cluster: 15] Attempt 1: Running complete annotation pipeline...\n",
      "250613:0721:56 |INFO| [Cluster: 14] Attempt 1: Running complete annotation pipeline...\n",
      "250613:0722:09 |INFO| [Cluster: 16] Attempt 1: Running complete annotation pipeline...\n",
      "250613:0722:13 |INFO| [Cluster: 18] Attempt 1: Running complete annotation pipeline...\n",
      "250613:0722:14 |INFO| [Cluster: 17] Attempt 1: Running complete annotation pipeline...\n",
      "250613:0722:53 |INFO| [Cluster: 14] Finding ontology term for annotation...\n",
      "250613:0723:05 |INFO| [Cluster: 16] Finding ontology term for annotation...\n",
      "250613:0723:05 |INFO| [Cluster: 15] Finding ontology term for annotation...\n",
      "250613:0723:21 |INFO| [Cluster: 18] Finding ontology term for annotation...\n",
      "250613:0723:22 |INFO| [Cluster: 17] Finding ontology term for annotation...\n",
      "250613:0723:51 |INFO| [Cluster: 15] Fetching CellGuide markers...\n",
      "250613:0723:51 |INFO| [Cluster: 15] Final review with CellGuide markers...\n",
      "250613:0723:52 |INFO| [Cluster: 16] Fetching CellGuide markers...\n",
      "250613:0723:52 |INFO| [Cluster: 16] Final review with CellGuide markers...\n",
      "250613:0724:01 |INFO| [Cluster: 14] Fetching CellGuide markers...\n",
      "250613:0724:01 |INFO| [Cluster: 14] Final review with CellGuide markers...\n",
      "250613:0724:12 |INFO| [Cluster: 18] Fetching CellGuide markers...\n",
      "250613:0724:12 |INFO| [Cluster: 18] Final review with CellGuide markers...\n",
      "250613:0724:28 |INFO| [Cluster: 17] Fetching CellGuide markers...\n",
      "250613:0724:28 |INFO| [Cluster: 17] Final review with CellGuide markers...\n",
      "250613:0724:56 |INFO| [Cluster: 16] Annotation not approved, will retry with feedback: Reviewer agent failed due to error: status_code: 403, model_name: qwen/qwen3-235b-a22b, body: {'mess...\n",
      "250613:0724:56 |INFO| [Cluster: 16] Retry 2: Re-running complete pipeline with feedback\n",
      "250613:0724:58 |INFO| [Cluster: 15] Annotation not approved, will retry with feedback: Reviewer agent failed due to error: status_code: 403, model_name: qwen/qwen3-235b-a22b, body: {'mess...\n",
      "250613:0724:58 |INFO| [Cluster: 15] Retry 2: Re-running complete pipeline with feedback\n",
      "250613:0725:10 |INFO| [Cluster: 18] Annotation not approved, will retry with feedback: Reviewer agent failed due to error: status_code: 403, model_name: qwen/qwen3-235b-a22b, body: {'mess...\n",
      "250613:0725:10 |INFO| [Cluster: 18] Retry 2: Re-running complete pipeline with feedback\n",
      "250613:0725:16 |INFO| [Cluster: 14] Annotation not approved, will retry with feedback: Reviewer agent failed due to error: status_code: 403, model_name: qwen/qwen3-235b-a22b, body: {'mess...\n",
      "250613:0725:16 |INFO| [Cluster: 14] Retry 2: Re-running complete pipeline with feedback\n",
      "250613:0725:42 |INFO| [Cluster: 17] Annotation not approved, will retry with feedback: Reviewer agent failed due to error: status_code: 403, model_name: qwen/qwen3-235b-a22b, body: {'mess...\n",
      "250613:0725:42 |INFO| [Cluster: 17] Retry 2: Re-running complete pipeline with feedback\n",
      "250613:0726:04 |INFO| [Cluster: 15] Finding ontology term for annotation...\n",
      "250613:0726:07 |INFO| [Cluster: 18] Finding ontology term for annotation...\n",
      "250613:0726:08 |INFO| [Cluster: 16] Finding ontology term for annotation...\n",
      "250613:0726:29 |INFO| [Cluster: 14] Finding ontology term for annotation...\n",
      "250613:0726:45 |INFO| [Cluster: 17] Finding ontology term for annotation...\n",
      "250613:0726:56 |INFO| [Cluster: 15] Fetching CellGuide markers...\n",
      "250613:0726:56 |INFO| [Cluster: 15] Final review with CellGuide markers...\n",
      "250613:0727:03 |INFO| [Cluster: 18] Fetching CellGuide markers...\n",
      "250613:0727:03 |INFO| [Cluster: 18] Final review with CellGuide markers...\n",
      "250613:0727:21 |INFO| [Cluster: 16] Fetching CellGuide markers...\n",
      "250613:0727:21 |INFO| [Cluster: 16] Final review with CellGuide markers...\n"
     ]
    },
    {
     "ename": "CyteTypeJobError",
     "evalue": "Server error: job failed",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mCyteTypeJobError\u001b[39m                          Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[6]\u001b[39m\u001b[32m, line 3\u001b[39m\n\u001b[32m      1\u001b[39m \u001b[38;5;66;03m# Zhang\u001b[39;00m\n\u001b[32m      2\u001b[39m \u001b[38;5;66;03m# Run annotation with custom model\u001b[39;00m\n\u001b[32m----> \u001b[39m\u001b[32m3\u001b[39m adata = \u001b[43mannotator\u001b[49m\u001b[43m.\u001b[49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m      4\u001b[39m \u001b[43m    \u001b[49m\u001b[43mstudy_context\u001b[49m\u001b[43m \u001b[49m\u001b[43m=\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mHealthy adult human bone marrow, analysed using 10x genomics single cell RNA-seq kits\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m      5\u001b[39m \u001b[43m    \u001b[49m\u001b[43mmodel_config\u001b[49m\u001b[43m \u001b[49m\u001b[43m=\u001b[49m\u001b[43m \u001b[49m\u001b[43m[\u001b[49m\u001b[43m{\u001b[49m\n\u001b[32m      6\u001b[39m \u001b[43m        \u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mprovider\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mopenai\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m      7\u001b[39m \u001b[43m        \u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mname\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mqwen/qwen3-235b-a22b\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m      8\u001b[39m \u001b[43m        \u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mapiKey\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mos\u001b[49m\u001b[43m.\u001b[49m\u001b[43menviron\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mOPENROUTER_API_KEY\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m      9\u001b[39m \u001b[43m        \u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mbaseUrl\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mhttps://openrouter.ai/api/v1\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m     10\u001b[39m \u001b[43m    \u001b[49m\u001b[43m}\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m     11\u001b[39m \u001b[43m    \u001b[49m\u001b[43mauth_token\u001b[49m\u001b[43m \u001b[49m\u001b[43m=\u001b[49m\u001b[43m \u001b[49m\u001b[43mos\u001b[49m\u001b[43m.\u001b[49m\u001b[43menviron\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mNYGEN_API_KEY\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m     12\u001b[39m \u001b[43m    \u001b[49m\u001b[43mtimeout_seconds\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m3900\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m     13\u001b[39m \u001b[43m    \u001b[49m\u001b[43mrun_config\u001b[49m\u001b[43m=\u001b[49m\u001b[43m{\u001b[49m\n\u001b[32m     14\u001b[39m \u001b[43m        \u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mconcurrentClusters\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[32;43m5\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m        \u001b[49m\u001b[38;5;66;43;03m# Default: 3, Range: 2-10\u001b[39;49;00m\n\u001b[32m     15\u001b[39m \u001b[43m        \u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mmaxAnnotationRevisions\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[32;43m2\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m    \u001b[49m\u001b[38;5;66;43;03m# Default: 2, Range: 1-5\u001b[39;49;00m\n\u001b[32m     16\u001b[39m \u001b[43m        \u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mmaxLLMRequests\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[32;43m900\u001b[39;49m\u001b[43m           \u001b[49m\u001b[38;5;66;43;03m# Default: 500, Range: 50-2000\u001b[39;49;00m\n\u001b[32m     17\u001b[39m \u001b[43m    \u001b[49m\u001b[43m}\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m     18\u001b[39m \u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/benchLLM/lib/python3.12/site-packages/cytetype/main.py:249\u001b[39m, in \u001b[36mCyteType.run\u001b[39m\u001b[34m(self, study_context, model_config, run_config, results_prefix, poll_interval_seconds, timeout_seconds, api_url, save_query, auth_token)\u001b[39m\n\u001b[32m    244\u001b[39m report_url = \u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mapi_url\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m/report/\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mjob_id\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m\"\u001b[39m\n\u001b[32m    245\u001b[39m logger.info(\n\u001b[32m    246\u001b[39m     \u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mView the automatically updating visualization report at: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mreport_url\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m\"\u001b[39m\n\u001b[32m    247\u001b[39m )\n\u001b[32m--> \u001b[39m\u001b[32m249\u001b[39m result = \u001b[43mpoll_for_results\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    250\u001b[39m \u001b[43m    \u001b[49m\u001b[43mjob_id\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    251\u001b[39m \u001b[43m    \u001b[49m\u001b[43mapi_url\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    252\u001b[39m \u001b[43m    \u001b[49m\u001b[43mpoll_interval_seconds\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    253\u001b[39m \u001b[43m    \u001b[49m\u001b[43mtimeout_seconds\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    254\u001b[39m \u001b[43m    \u001b[49m\u001b[43mauth_token\u001b[49m\u001b[43m=\u001b[49m\u001b[43mauth_token\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    255\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    257\u001b[39m \u001b[38;5;66;03m# Store results in AnnData object\u001b[39;00m\n\u001b[32m    258\u001b[39m \u001b[38;5;28mself\u001b[39m.adata.uns[\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mresults_prefix\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m_results\u001b[39m\u001b[33m\"\u001b[39m] = {\n\u001b[32m    259\u001b[39m     \u001b[33m\"\u001b[39m\u001b[33mjob_id\u001b[39m\u001b[33m\"\u001b[39m: job_id,\n\u001b[32m    260\u001b[39m     \u001b[33m\"\u001b[39m\u001b[33mresult\u001b[39m\u001b[33m\"\u001b[39m: result,\n\u001b[32m    261\u001b[39m }\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/benchLLM/lib/python3.12/site-packages/cytetype/client.py:131\u001b[39m, in \u001b[36mpoll_for_results\u001b[39m\u001b[34m(job_id, api_url, poll_interval, timeout, auth_token)\u001b[39m\n\u001b[32m    129\u001b[39m     error_message = data.get(\u001b[33m\"\u001b[39m\u001b[33mmessage\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mUnknown error\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m    130\u001b[39m     logger.debug(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mJob \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mjob_id\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m failed on the server: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00merror_message\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m\"\u001b[39m)\n\u001b[32m--> \u001b[39m\u001b[32m131\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m CyteTypeJobError(\u001b[33m\"\u001b[39m\u001b[33mServer error: job failed\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m    132\u001b[39m \u001b[38;5;28;01melif\u001b[39;00m status \u001b[38;5;129;01min\u001b[39;00m [\u001b[33m\"\u001b[39m\u001b[33mprocessing\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mpending\u001b[39m\u001b[33m\"\u001b[39m]:\n\u001b[32m    133\u001b[39m     logger.debug(\n\u001b[32m    134\u001b[39m         \u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mJob \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mjob_id\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m status: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mstatus\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m. Checking logs and waiting \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mpoll_interval\u001b[38;5;132;01m}\u001b[39;00m\u001b[33ms...\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m    135\u001b[39m     )\n",
      "\u001b[31mCyteTypeJobError\u001b[39m: Server error: job failed"
     ]
    }
   ],
   "source": [
    "# Zhang\n",
    "# Run annotation with custom model\n",
    "adata = annotator.run(\n",
    "    study_context = \"Healthy adult human bone marrow, analysed using 10x genomics single cell RNA-seq kits\",\n",
    "    model_config = [{\n",
    "        'provider': 'openai',\n",
    "        'name': 'qwen/qwen3-235b-a22b',\n",
    "        'apiKey': os.environ.get('OPENROUTER_API_KEY'),\n",
    "        'baseUrl': \"https://openrouter.ai/api/v1\",\n",
    "    }],\n",
    "    auth_token = os.environ.get('NYGEN_API_KEY'),\n",
    "    timeout_seconds=3900,\n",
    "    run_config={\n",
    "        'concurrentClusters': 5,        # Default: 3, Range: 2-10\n",
    "        'maxAnnotationRevisions': 2,    # Default: 2, Range: 1-5\n",
    "        'maxLLMRequests': 900           # Default: 500, Range: 50-2000\n",
    "    },\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "92da50e9-814e-4a18-82c9-e0e62e442246",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true,
     "source_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Waiting for results for job ID: 9b2e9900-7410-4380-84ac-ccefcdbe2498\n",
      "View the automatically updating visualization report at: https://nygen-labs-prod--cell-annotation-agent-fastapi-app.modal.run/report/9b2e9900-7410-4380-84ac-ccefcdbe2498\n",
      "250610:1300:28 |INFO| [STARTING WORKFLOWS]: Job ID: 9b2e9900-7410-4380-84ac-ccefcdbe2498\n",
      "250610:1300:28 |INFO| [WORKFLOW STEP 1/3]: Generating context summary\n",
      "250610:1300:51 |INFO| [WORKFLOW STEP 2/3]: Running complete annotation pipeline for each cluster\n",
      "250610:1300:52 |INFO| [Cluster: 1] Generating cluster-specific context...\n",
      "250610:1300:52 |INFO| [Cluster: 10] Generating cluster-specific context...\n",
      "250610:1300:53 |INFO| [Cluster: 11] Generating cluster-specific context...\n",
      "250610:1300:55 |INFO| [Cluster: 12] Generating cluster-specific context...\n",
      "250610:1300:56 |INFO| [Cluster: 13] Generating cluster-specific context...\n",
      "250610:1301:02 |INFO| [Cluster: 10] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1301:02 |INFO| [Cluster: 11] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1301:02 |INFO| [Cluster: 12] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1301:02 |INFO| [Cluster: 1] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1301:03 |INFO| [Cluster: 13] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1301:36 |INFO| [Cluster: 1] Finding ontology term for annotation...\n",
      "250610:1301:41 |INFO| [Cluster: 12] Finding ontology term for annotation...\n",
      "250610:1301:43 |INFO| [Cluster: 13] Finding ontology term for annotation...\n",
      "250610:1301:52 |INFO| [Cluster: 1] Fetching CellGuide markers...\n",
      "250610:1301:53 |INFO| [Cluster: 10] Finding ontology term for annotation...\n",
      "250610:1302:15 |INFO| [Cluster: 13] Fetching CellGuide markers...\n",
      "250610:1302:16 |INFO| [Cluster: 10] Fetching CellGuide markers...\n",
      "250610:1302:16 |INFO| [Cluster: 10] No CellGuide canonical markers found for CL_0002355\n",
      "250610:1302:22 |INFO| [Cluster: 12] Fetching CellGuide markers...\n",
      "250610:1302:22 |INFO| [Cluster: 10] No CellGuide canonical markers found for CL_0002355\n",
      "250610:1302:33 |INFO| [Cluster: 10] No CellGuide canonical markers found for CL_0002355\n",
      "250610:1302:38 |INFO| [Cluster: 11] Finding ontology term for annotation...\n",
      "250610:1302:47 |INFO| [Cluster: 1] Final review with CellGuide markers...\n",
      "250610:1302:50 |INFO| [Cluster: 11] Fetching CellGuide markers...\n",
      "250610:1303:14 |INFO| [Cluster: 13] Final review with CellGuide markers...\n",
      "250610:1303:22 |INFO| [Cluster: 10] No CellGuide canonical markers found for CL_0002355\n",
      "250610:1303:22 |INFO| [Cluster: 1] Annotation approved by reviewer\n",
      "250610:1303:22 |SUCCESS| [Cluster: 1] Annotation pipeline completed in 150.23s after 1 attempt(s)\n",
      "250610:1303:22 |INFO| [Cluster: 1] LLM Usage: 10 requests, 52752 input tokens, 1220 output tokens, 53972 total tokens\n",
      "250610:1303:22 |INFO| [Cluster: 1] Saving partial result (1/21)\n",
      "250610:1303:24 |INFO| [Cluster: 14] Generating cluster-specific context...\n",
      "250610:1303:24 |INFO| [Cluster: 10] Final review with CellGuide markers...\n",
      "250610:1303:28 |INFO| [Cluster: 12] Final review with CellGuide markers...\n",
      "250610:1303:34 |INFO| [Cluster: 14] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1303:58 |INFO| [Cluster: 12] Annotation approved by reviewer\n",
      "250610:1303:58 |SUCCESS| [Cluster: 12] Annotation pipeline completed in 183.45s after 1 attempt(s)\n",
      "250610:1303:58 |INFO| [Cluster: 12] LLM Usage: 14 requests, 73321 input tokens, 1367 output tokens, 74688 total tokens\n",
      "250610:1303:58 |INFO| [Cluster: 12] Saving partial result (2/21)\n",
      "250610:1304:00 |INFO| [Cluster: 15] Generating cluster-specific context...\n",
      "250610:1304:00 |INFO| [Cluster: 13] Annotation not approved, will retry with feedback: Re-annotate as 'early B cell precursor (pro-B/pre-B cell)' based on the overwhelming presence of B-l...\n",
      "250610:1304:00 |INFO| [Cluster: 13] Retry 2: Re-running complete pipeline with feedback\n",
      "250610:1304:03 |INFO| [Cluster: 11] Final review with CellGuide markers...\n",
      "250610:1304:07 |INFO| [Cluster: 10] Annotation approved by reviewer\n",
      "250610:1304:07 |SUCCESS| [Cluster: 10] Annotation pipeline completed in 194.92s after 1 attempt(s)\n",
      "250610:1304:07 |INFO| [Cluster: 10] LLM Usage: 10 requests, 51439 input tokens, 1366 output tokens, 52805 total tokens\n",
      "250610:1304:07 |INFO| [Cluster: 10] Saving partial result (3/21)\n",
      "250610:1304:09 |INFO| [Cluster: 16] Generating cluster-specific context...\n",
      "250610:1304:11 |INFO| [Cluster: 15] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1304:17 |INFO| [Cluster: 16] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1304:28 |INFO| [Cluster: 14] Finding ontology term for annotation...\n",
      "250610:1304:39 |INFO| [Cluster: 14] Fetching CellGuide markers...\n",
      "250610:1304:39 |INFO| [Cluster: 14] No CellGuide canonical markers found for CL_0002353\n",
      "250610:1304:39 |INFO| [Cluster: 14] No CellGuide computational markers found for CL_0002353. Data: {'error': 'No JSON data for computational markers for CL_0002353.'}\n",
      "250610:1304:41 |INFO| [Cluster: 13] Finding ontology term for annotation...\n",
      "250610:1304:44 |INFO| [Cluster: 11] Annotation not approved, will retry with feedback: Revise the annotation to 'B cell precursor' (granular: 'pro-B/pre-B cell, CD79A+, IGLL1+, VPREB1+, D...\n",
      "250610:1304:44 |INFO| [Cluster: 11] Retry 2: Re-running complete pipeline with feedback\n",
      "250610:1304:44 |INFO| [Cluster: 14] No CellGuide canonical markers found for CL_0002353\n",
      "250610:1304:44 |INFO| [Cluster: 14] No CellGuide computational markers found for CL_0002353. Data: {'error': 'No JSON data for computational markers for CL_0002353.'}\n",
      "250610:1304:44 |INFO| [Cluster: 13] Fetching CellGuide markers...\n",
      "250610:1304:45 |INFO| [Cluster: 13] No CellGuide canonical markers found for CL_0002046\n",
      "250610:1304:50 |INFO| [Cluster: 13] No CellGuide canonical markers found for CL_0002046\n",
      "250610:1304:52 |INFO| [Cluster: 15] Finding ontology term for annotation...\n",
      "250610:1304:55 |INFO| [Cluster: 14] No CellGuide canonical markers found for CL_0002353\n",
      "250610:1304:55 |INFO| [Cluster: 14] No CellGuide computational markers found for CL_0002353. Data: {'error': 'No JSON data for computational markers for CL_0002353.'}\n",
      "250610:1305:01 |INFO| [Cluster: 13] No CellGuide canonical markers found for CL_0002046\n",
      "250610:1305:08 |INFO| [Cluster: 15] Fetching CellGuide markers...\n",
      "250610:1305:20 |INFO| [Cluster: 16] Finding ontology term for annotation...\n",
      "250610:1305:24 |INFO| [Cluster: 11] Finding ontology term for annotation...\n",
      "250610:1305:38 |INFO| [Cluster: 16] Fetching CellGuide markers...\n",
      "250610:1305:38 |INFO| [Cluster: 16] No CellGuide canonical markers found for CL_0002353\n",
      "250610:1305:38 |INFO| [Cluster: 16] No CellGuide computational markers found for CL_0002353. Data: {'error': 'No JSON data for computational markers for CL_0002353.'}\n",
      "250610:1305:39 |INFO| [Cluster: 14] No CellGuide canonical markers found for CL_0002353\n",
      "250610:1305:39 |INFO| [Cluster: 14] No CellGuide computational markers found for CL_0002353. Data: {'error': 'No JSON data for computational markers for CL_0002353.'}\n",
      "250610:1305:39 |INFO| [Cluster: 11] Fetching CellGuide markers...\n",
      "250610:1305:39 |INFO| [Cluster: 14] Final review with CellGuide markers...\n",
      "250610:1305:45 |INFO| [Cluster: 16] No CellGuide canonical markers found for CL_0002353\n",
      "250610:1305:45 |INFO| [Cluster: 16] No CellGuide computational markers found for CL_0002353. Data: {'error': 'No JSON data for computational markers for CL_0002353.'}\n",
      "250610:1305:55 |INFO| [Cluster: 14] Annotation approved by reviewer\n",
      "250610:1305:55 |SUCCESS| [Cluster: 14] Annotation pipeline completed in 151.40s after 1 attempt(s)\n",
      "250610:1305:55 |INFO| [Cluster: 14] LLM Usage: 13 requests, 63953 input tokens, 1321 output tokens, 65274 total tokens\n",
      "250610:1305:55 |INFO| [Cluster: 14] Saving partial result (4/21)\n",
      "250610:1305:57 |INFO| [Cluster: 17] Generating cluster-specific context...\n",
      "250610:1305:57 |INFO| [Cluster: 13] No CellGuide canonical markers found for CL_0002046\n",
      "250610:1305:57 |INFO| [Cluster: 16] No CellGuide canonical markers found for CL_0002353\n",
      "250610:1305:57 |INFO| [Cluster: 16] No CellGuide computational markers found for CL_0002353. Data: {'error': 'No JSON data for computational markers for CL_0002353.'}\n",
      "250610:1305:58 |INFO| [Cluster: 13] Final review with CellGuide markers...\n",
      "250610:1306:05 |INFO| [Cluster: 17] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1306:16 |INFO| [Cluster: 13] Reached max retries (2), proceeding with current annotation\n",
      "250610:1306:16 |SUCCESS| [Cluster: 13] Annotation pipeline completed in 320.58s after 2 attempt(s)\n",
      "250610:1306:16 |INFO| [Cluster: 13] LLM Usage: 20 requests, 99652 input tokens, 2387 output tokens, 102039 total tokens\n",
      "250610:1306:16 |INFO| [Cluster: 13] Saving partial result (5/21)\n",
      "250610:1306:19 |INFO| [Cluster: 18] Generating cluster-specific context...\n",
      "250610:1306:25 |INFO| [Cluster: 18] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1306:26 |INFO| [Cluster: 15] Final review with CellGuide markers...\n",
      "250610:1306:34 |INFO| [Cluster: 16] No CellGuide canonical markers found for CL_0002353\n",
      "250610:1306:34 |INFO| [Cluster: 16] No CellGuide computational markers found for CL_0002353. Data: {'error': 'No JSON data for computational markers for CL_0002353.'}\n",
      "250610:1306:35 |INFO| [Cluster: 16] Final review with CellGuide markers...\n",
      "250610:1306:41 |INFO| [Cluster: 11] Final review with CellGuide markers...\n",
      "250610:1306:55 |INFO| [Cluster: 17] Finding ontology term for annotation...\n",
      "250610:1306:58 |INFO| [Cluster: 15] Annotation not approved, will retry with feedback: Revise the annotation to B cell precursor (pro-B or pre-B cell), based on the dominant expression of...\n",
      "250610:1306:58 |INFO| [Cluster: 15] Retry 2: Re-running complete pipeline with feedback\n",
      "250610:1306:59 |INFO| [Cluster: 17] Fetching CellGuide markers...\n",
      "250610:1307:03 |INFO| [Cluster: 16] Annotation approved by reviewer\n",
      "250610:1307:03 |SUCCESS| [Cluster: 16] Annotation pipeline completed in 174.27s after 1 attempt(s)\n",
      "250610:1307:03 |INFO| [Cluster: 16] LLM Usage: 14 requests, 75135 input tokens, 1515 output tokens, 76650 total tokens\n",
      "250610:1307:03 |INFO| [Cluster: 16] Saving partial result (6/21)\n",
      "250610:1307:05 |INFO| [Cluster: 19] Generating cluster-specific context...\n",
      "250610:1307:07 |INFO| [Cluster: 11] Reached max retries (2), proceeding with current annotation\n",
      "250610:1307:07 |SUCCESS| [Cluster: 11] Annotation pipeline completed in 373.97s after 2 attempt(s)\n",
      "250610:1307:07 |INFO| [Cluster: 11] LLM Usage: 25 requests, 134451 input tokens, 2679 output tokens, 137130 total tokens\n",
      "250610:1307:07 |INFO| [Cluster: 11] Saving partial result (7/21)\n",
      "250610:1307:09 |INFO| [Cluster: 2] Generating cluster-specific context...\n",
      "250610:1307:13 |INFO| [Cluster: 19] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1307:20 |INFO| [Cluster: 2] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1307:28 |INFO| [Cluster: 18] Finding ontology term for annotation...\n",
      "250610:1307:39 |INFO| [Cluster: 18] Fetching CellGuide markers...\n",
      "250610:1307:40 |INFO| [Cluster: 18] No CellGuide canonical markers found for CL_4033081\n",
      "250610:1307:40 |INFO| [Cluster: 18] No CellGuide computational markers found for CL_4033081. Data: {'error': 'No JSON data for computational markers for CL_4033081.'}\n",
      "250610:1307:43 |INFO| [Cluster: 15] Finding ontology term for annotation...\n",
      "250610:1307:45 |INFO| [Cluster: 18] No CellGuide canonical markers found for CL_4033081\n",
      "250610:1307:45 |INFO| [Cluster: 18] No CellGuide computational markers found for CL_4033081. Data: {'error': 'No JSON data for computational markers for CL_4033081.'}\n",
      "250610:1307:56 |INFO| [Cluster: 18] No CellGuide canonical markers found for CL_4033081\n",
      "250610:1307:56 |INFO| [Cluster: 18] No CellGuide computational markers found for CL_4033081. Data: {'error': 'No JSON data for computational markers for CL_4033081.'}\n",
      "250610:1307:57 |INFO| [Cluster: 19] Finding ontology term for annotation...\n",
      "250610:1307:59 |INFO| [Cluster: 2] Finding ontology term for annotation...\n",
      "250610:1308:02 |INFO| [Cluster: 2] Fetching CellGuide markers...\n",
      "250610:1308:02 |INFO| [Cluster: 2] No CellGuide canonical markers found for CL_0000957\n",
      "250610:1308:02 |INFO| [Cluster: 15] Fetching CellGuide markers...\n",
      "250610:1308:08 |INFO| [Cluster: 2] No CellGuide canonical markers found for CL_0000957\n",
      "250610:1308:12 |INFO| [Cluster: 19] Fetching CellGuide markers...\n",
      "250610:1308:17 |INFO| [Cluster: 17] Final review with CellGuide markers...\n",
      "250610:1308:18 |INFO| [Cluster: 2] No CellGuide canonical markers found for CL_0000957\n",
      "250610:1308:37 |INFO| [Cluster: 17] Annotation not approved, will retry with feedback: Revise the annotation to reflect a B cell precursor (pro-B or pre-B cell) identity, based on the ove...\n",
      "250610:1308:37 |INFO| [Cluster: 17] Retry 2: Re-running complete pipeline with feedback\n",
      "250610:1308:44 |INFO| [Cluster: 18] No CellGuide canonical markers found for CL_4033081\n",
      "250610:1308:44 |INFO| [Cluster: 18] No CellGuide computational markers found for CL_4033081. Data: {'error': 'No JSON data for computational markers for CL_4033081.'}\n",
      "250610:1308:45 |INFO| [Cluster: 18] Final review with CellGuide markers...\n",
      "250610:1309:04 |INFO| [Cluster: 2] No CellGuide canonical markers found for CL_0000957\n",
      "250610:1309:04 |INFO| [Cluster: 2] Final review with CellGuide markers...\n",
      "250610:1309:06 |INFO| [Cluster: 18] Annotation not approved, will retry with feedback: Revise the annotation to reflect a cycling B cell precursor identity, as the dominant marker set is ...\n",
      "250610:1309:06 |INFO| [Cluster: 18] Retry 2: Re-running complete pipeline with feedback\n",
      "250610:1309:14 |INFO| [Cluster: 19] Final review with CellGuide markers...\n",
      "250610:1309:14 |INFO| [Cluster: 15] Final review with CellGuide markers...\n",
      "250610:1309:21 |INFO| [Cluster: 2] Annotation approved by reviewer\n",
      "250610:1309:21 |SUCCESS| [Cluster: 2] Annotation pipeline completed in 131.68s after 1 attempt(s)\n",
      "250610:1309:21 |INFO| [Cluster: 2] LLM Usage: 10 requests, 49576 input tokens, 1202 output tokens, 50778 total tokens\n",
      "250610:1309:21 |INFO| [Cluster: 2] Saving partial result (8/21)\n",
      "250610:1309:23 |INFO| [Cluster: 20] Generating cluster-specific context...\n",
      "250610:1309:29 |INFO| [Cluster: 17] Finding ontology term for annotation...\n",
      "250610:1309:37 |INFO| [Cluster: 19] Annotation not approved, will retry with feedback: Revise the annotation to reflect a B cell precursor identity, supported by the dominant expression o...\n",
      "250610:1309:37 |INFO| [Cluster: 19] Retry 2: Re-running complete pipeline with feedback\n",
      "250610:1309:40 |INFO| [Cluster: 20] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1309:45 |INFO| [Cluster: 18] Finding ontology term for annotation...\n",
      "250610:1309:46 |INFO| [Cluster: 15] Reached max retries (2), proceeding with current annotation\n",
      "250610:1309:46 |SUCCESS| [Cluster: 15] Annotation pipeline completed in 346.57s after 2 attempt(s)\n",
      "250610:1309:46 |INFO| [Cluster: 15] LLM Usage: 24 requests, 133282 input tokens, 2764 output tokens, 136046 total tokens\n",
      "250610:1309:46 |INFO| [Cluster: 15] Saving partial result (9/21)\n",
      "250610:1309:48 |INFO| [Cluster: 21] Generating cluster-specific context...\n",
      "250610:1309:52 |INFO| [Cluster: 17] Fetching CellGuide markers...\n",
      "250610:1309:53 |INFO| [Cluster: 17] No CellGuide canonical markers found for CL_0000776\n",
      "250610:1309:58 |INFO| [Cluster: 17] No CellGuide canonical markers found for CL_0000776\n",
      "250610:1310:01 |INFO| [Cluster: 21] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1310:01 |INFO| [Cluster: 18] Fetching CellGuide markers...\n",
      "250610:1310:01 |INFO| [Cluster: 18] No CellGuide canonical markers found for CL_0000957\n",
      "250610:1310:07 |INFO| [Cluster: 18] No CellGuide canonical markers found for CL_0000957\n",
      "250610:1310:09 |INFO| [Cluster: 17] No CellGuide canonical markers found for CL_0000776\n",
      "250610:1310:18 |INFO| [Cluster: 18] No CellGuide canonical markers found for CL_0000957\n",
      "250610:1310:26 |INFO| [Cluster: 20] Finding ontology term for annotation...\n",
      "250610:1310:34 |INFO| [Cluster: 19] Finding ontology term for annotation...\n",
      "250610:1310:39 |INFO| [Cluster: 20] Fetching CellGuide markers...\n",
      "250610:1310:39 |INFO| [Cluster: 21] Finding ontology term for annotation...\n",
      "250610:1310:43 |INFO| [Cluster: 21] Fetching CellGuide markers...\n",
      "250610:1310:49 |INFO| [Cluster: 19] Fetching CellGuide markers...\n",
      "250610:1311:01 |INFO| [Cluster: 18] No CellGuide canonical markers found for CL_0000957\n",
      "250610:1311:04 |INFO| [Cluster: 17] No CellGuide canonical markers found for CL_0000776\n",
      "250610:1311:05 |INFO| [Cluster: 17] Final review with CellGuide markers...\n",
      "250610:1311:08 |INFO| [Cluster: 18] Final review with CellGuide markers...\n",
      "250610:1311:24 |INFO| [Cluster: 18] Reached max retries (2), proceeding with current annotation\n",
      "250610:1311:24 |SUCCESS| [Cluster: 18] Annotation pipeline completed in 304.90s after 2 attempt(s)\n",
      "250610:1311:24 |INFO| [Cluster: 18] LLM Usage: 26 requests, 158649 input tokens, 2825 output tokens, 161474 total tokens\n",
      "250610:1311:24 |INFO| [Cluster: 18] Saving partial result (10/21)\n",
      "250610:1311:27 |INFO| [Cluster: 3] Generating cluster-specific context...\n",
      "250610:1311:27 |INFO| [Cluster: 17] Reached max retries (2), proceeding with current annotation\n",
      "250610:1311:27 |SUCCESS| [Cluster: 17] Annotation pipeline completed in 329.17s after 2 attempt(s)\n",
      "250610:1311:27 |INFO| [Cluster: 17] LLM Usage: 22 requests, 119324 input tokens, 2732 output tokens, 122056 total tokens\n",
      "250610:1311:27 |INFO| [Cluster: 17] Saving partial result (11/21)\n",
      "250610:1311:29 |INFO| [Cluster: 4] Generating cluster-specific context...\n",
      "250610:1311:34 |INFO| [Cluster: 3] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1311:37 |INFO| [Cluster: 4] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1311:42 |INFO| [Cluster: 21] Final review with CellGuide markers...\n",
      "250610:1311:52 |INFO| [Cluster: 19] Final review with CellGuide markers...\n",
      "250610:1311:54 |INFO| [Cluster: 20] Final review with CellGuide markers...\n",
      "250610:1312:05 |INFO| [Cluster: 19] Reached max retries (2), proceeding with current annotation\n",
      "250610:1312:05 |SUCCESS| [Cluster: 19] Annotation pipeline completed in 299.61s after 2 attempt(s)\n",
      "250610:1312:05 |INFO| [Cluster: 19] LLM Usage: 26 requests, 148460 input tokens, 2798 output tokens, 151258 total tokens\n",
      "250610:1312:05 |INFO| [Cluster: 19] Saving partial result (12/21)\n",
      "250610:1312:07 |INFO| [Cluster: 5] Generating cluster-specific context...\n",
      "250610:1312:16 |INFO| [Cluster: 21] Annotation not approved, will retry with feedback: Revise the annotation to 'Proliferating B cell precursor (pre-B/pro-B)' as the primary identity, wit...\n",
      "250610:1312:16 |INFO| [Cluster: 21] Retry 2: Re-running complete pipeline with feedback\n",
      "250610:1312:16 |INFO| [Cluster: 5] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1312:17 |INFO| [Cluster: 3] Finding ontology term for annotation...\n",
      "250610:1312:24 |INFO| [Cluster: 3] Fetching CellGuide markers...\n",
      "250610:1312:25 |INFO| [Cluster: 4] Finding ontology term for annotation...\n",
      "250610:1312:34 |INFO| [Cluster: 20] Annotation not approved, will retry with feedback: Revise the annotation to reflect a B cell precursor identity, supported by the dominant expression o...\n",
      "250610:1312:34 |INFO| [Cluster: 20] Retry 2: Re-running complete pipeline with feedback\n",
      "250610:1312:54 |INFO| [Cluster: 4] Fetching CellGuide markers...\n",
      "250610:1313:03 |INFO| [Cluster: 5] Finding ontology term for annotation...\n",
      "250610:1313:10 |INFO| [Cluster: 5] Fetching CellGuide markers...\n",
      "250610:1313:10 |INFO| [Cluster: 5] No CellGuide canonical markers found for CL_1000398\n",
      "250610:1313:15 |INFO| [Cluster: 5] No CellGuide canonical markers found for CL_1000398\n",
      "250610:1313:18 |INFO| [Cluster: 21] Finding ontology term for annotation...\n",
      "250610:1313:26 |INFO| [Cluster: 5] No CellGuide canonical markers found for CL_1000398\n",
      "250610:1313:37 |INFO| [Cluster: 20] Finding ontology term for annotation...\n",
      "250610:1313:39 |INFO| [Cluster: 3] Final review with CellGuide markers...\n",
      "250610:1313:42 |INFO| [Cluster: 21] Fetching CellGuide markers...\n",
      "250610:1313:42 |INFO| [Cluster: 21] No CellGuide canonical markers found for CL_4033068\n",
      "250610:1313:42 |INFO| [Cluster: 21] No CellGuide computational markers found for CL_4033068. Data: {'error': 'No JSON data for computational markers for CL_4033068.'}\n",
      "250610:1313:48 |INFO| [Cluster: 21] No CellGuide canonical markers found for CL_4033068\n",
      "250610:1313:48 |INFO| [Cluster: 21] No CellGuide computational markers found for CL_4033068. Data: {'error': 'No JSON data for computational markers for CL_4033068.'}\n",
      "250610:1313:57 |INFO| [Cluster: 20] Fetching CellGuide markers...\n",
      "250610:1313:58 |INFO| [Cluster: 21] No CellGuide canonical markers found for CL_4033068\n",
      "250610:1313:58 |INFO| [Cluster: 21] No CellGuide computational markers found for CL_4033068. Data: {'error': 'No JSON data for computational markers for CL_4033068.'}\n",
      "250610:1314:02 |INFO| [Cluster: 3] Annotation approved by reviewer\n",
      "250610:1314:02 |SUCCESS| [Cluster: 3] Annotation pipeline completed in 155.79s after 1 attempt(s)\n",
      "250610:1314:02 |INFO| [Cluster: 3] LLM Usage: 10 requests, 53595 input tokens, 1342 output tokens, 54937 total tokens\n",
      "250610:1314:02 |INFO| [Cluster: 3] Saving partial result (13/21)\n",
      "250610:1314:05 |INFO| [Cluster: 6] Generating cluster-specific context...\n",
      "250610:1314:06 |INFO| [Cluster: 4] Final review with CellGuide markers...\n",
      "250610:1314:13 |INFO| [Cluster: 6] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1314:18 |INFO| [Cluster: 5] No CellGuide canonical markers found for CL_1000398\n",
      "250610:1314:18 |INFO| [Cluster: 5] Final review with CellGuide markers...\n",
      "250610:1314:33 |INFO| [Cluster: 21] No CellGuide canonical markers found for CL_4033068\n",
      "250610:1314:33 |INFO| [Cluster: 21] No CellGuide computational markers found for CL_4033068. Data: {'error': 'No JSON data for computational markers for CL_4033068.'}\n",
      "250610:1314:34 |INFO| [Cluster: 21] Final review with CellGuide markers...\n",
      "250610:1314:36 |INFO| [Cluster: 5] Annotation not approved, will retry with feedback: Revise annotation to 'B cell precursor' (pro-B or pre-B cell) based on the dominant B-lineage marker...\n",
      "250610:1314:36 |INFO| [Cluster: 5] Retry 2: Re-running complete pipeline with feedback\n",
      "250610:1314:51 |INFO| [Cluster: 20] Final review with CellGuide markers...\n",
      "250610:1314:59 |INFO| [Cluster: 4] Annotation not approved, will retry with feedback: Revise the annotation to reflect the strong early B cell precursor signature (pro-B/pre-B), and addr...\n",
      "250610:1314:59 |INFO| [Cluster: 4] Retry 2: Re-running complete pipeline with feedback\n",
      "250610:1315:00 |INFO| [Cluster: 6] Finding ontology term for annotation...\n",
      "250610:1315:06 |INFO| [Cluster: 21] Reached max retries (2), proceeding with current annotation\n",
      "250610:1315:06 |SUCCESS| [Cluster: 21] Annotation pipeline completed in 317.32s after 2 attempt(s)\n",
      "250610:1315:06 |INFO| [Cluster: 21] LLM Usage: 22 requests, 120126 input tokens, 2699 output tokens, 122825 total tokens\n",
      "250610:1315:06 |INFO| [Cluster: 21] Saving partial result (14/21)\n",
      "250610:1315:08 |INFO| [Cluster: 7] Generating cluster-specific context...\n",
      "250610:1315:09 |INFO| [Cluster: 6] Fetching CellGuide markers...\n",
      "250610:1315:30 |INFO| [Cluster: 20] Reached max retries (2), proceeding with current annotation\n",
      "250610:1315:30 |SUCCESS| [Cluster: 20] Annotation pipeline completed in 366.81s after 2 attempt(s)\n",
      "250610:1315:30 |INFO| [Cluster: 20] LLM Usage: 22 requests, 128374 input tokens, 2909 output tokens, 131283 total tokens\n",
      "250610:1315:30 |INFO| [Cluster: 20] Saving partial result (15/21)\n",
      "250610:1315:32 |INFO| [Cluster: 8] Generating cluster-specific context...\n",
      "250610:1315:38 |INFO| [Cluster: 8] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1315:51 |INFO| [Cluster: 4] Finding ontology term for annotation...\n",
      "250610:1315:54 |INFO| [Cluster: 7] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1316:04 |INFO| [Cluster: 5] Finding ontology term for annotation...\n",
      "250610:1316:05 |INFO| [Cluster: 4] Fetching CellGuide markers...\n",
      "250610:1316:06 |INFO| [Cluster: 4] No CellGuide canonical markers found for CL_0002046\n",
      "250610:1316:10 |INFO| [Cluster: 5] Fetching CellGuide markers...\n",
      "250610:1316:10 |INFO| [Cluster: 5] No CellGuide canonical markers found for CL_1000398\n",
      "250610:1316:11 |INFO| [Cluster: 4] No CellGuide canonical markers found for CL_0002046\n",
      "250610:1316:14 |INFO| [Cluster: 6] Final review with CellGuide markers...\n",
      "250610:1316:16 |INFO| [Cluster: 5] No CellGuide canonical markers found for CL_1000398\n",
      "250610:1316:22 |INFO| [Cluster: 4] No CellGuide canonical markers found for CL_0002046\n",
      "250610:1316:26 |INFO| [Cluster: 5] No CellGuide canonical markers found for CL_1000398\n",
      "250610:1316:27 |INFO| [Cluster: 8] Finding ontology term for annotation...\n",
      "250610:1316:30 |INFO| [Cluster: 6] Annotation approved by reviewer\n",
      "250610:1316:30 |SUCCESS| [Cluster: 6] Annotation pipeline completed in 145.18s after 1 attempt(s)\n",
      "250610:1316:30 |INFO| [Cluster: 6] LLM Usage: 10 requests, 52175 input tokens, 1224 output tokens, 53399 total tokens\n",
      "250610:1316:30 |INFO| [Cluster: 6] Saving partial result (16/21)\n",
      "250610:1316:32 |INFO| [Cluster: 9] Generating cluster-specific context...\n",
      "250610:1316:34 |INFO| [Cluster: 7] Finding ontology term for annotation...\n",
      "250610:1316:35 |INFO| [Cluster: 8] Fetching CellGuide markers...\n",
      "250610:1316:36 |INFO| [Cluster: 8] No CellGuide canonical markers found for CL_0000038\n",
      "250610:1316:38 |INFO| [Cluster: 7] Fetching CellGuide markers...\n",
      "250610:1316:38 |INFO| [Cluster: 7] No CellGuide canonical markers found for CL_0000038\n",
      "250610:1316:41 |INFO| [Cluster: 8] No CellGuide canonical markers found for CL_0000038\n",
      "250610:1316:44 |INFO| [Cluster: 9] Attempt 1: Running complete annotation pipeline...\n",
      "250610:1316:45 |INFO| [Cluster: 7] No CellGuide canonical markers found for CL_0000038\n",
      "250610:1316:51 |INFO| [Cluster: 8] No CellGuide canonical markers found for CL_0000038\n",
      "250610:1316:53 |INFO| [Cluster: 4] No CellGuide canonical markers found for CL_0002046\n",
      "250610:1316:53 |INFO| [Cluster: 4] Final review with CellGuide markers...\n",
      "250610:1316:55 |INFO| [Cluster: 7] No CellGuide canonical markers found for CL_0000038\n",
      "250610:1317:14 |INFO| [Cluster: 5] No CellGuide canonical markers found for CL_1000398\n",
      "250610:1317:15 |INFO| [Cluster: 5] Final review with CellGuide markers...\n",
      "250610:1317:24 |INFO| [Cluster: 4] Reached max retries (2), proceeding with current annotation\n",
      "250610:1317:24 |SUCCESS| [Cluster: 4] Annotation pipeline completed in 355.65s after 2 attempt(s)\n",
      "250610:1317:24 |INFO| [Cluster: 4] LLM Usage: 26 requests, 137454 input tokens, 2798 output tokens, 140252 total tokens\n",
      "250610:1317:24 |INFO| [Cluster: 4] Saving partial result (17/21)\n",
      "250610:1317:31 |INFO| [Cluster: 9] Finding ontology term for annotation...\n",
      "250610:1317:39 |INFO| [Cluster: 7] No CellGuide canonical markers found for CL_0000038\n",
      "250610:1317:40 |INFO| [Cluster: 7] Final review with CellGuide markers...\n",
      "250610:1317:49 |INFO| [Cluster: 8] No CellGuide canonical markers found for CL_0000038\n",
      "250610:1317:49 |INFO| [Cluster: 8] Final review with CellGuide markers...\n",
      "250610:1317:50 |INFO| [Cluster: 5] Reached max retries (2), proceeding with current annotation\n",
      "250610:1317:50 |SUCCESS| [Cluster: 5] Annotation pipeline completed in 343.25s after 2 attempt(s)\n",
      "250610:1317:50 |INFO| [Cluster: 5] LLM Usage: 19 requests, 128948 input tokens, 2533 output tokens, 131481 total tokens\n",
      "250610:1317:50 |INFO| [Cluster: 5] Saving partial result (18/21)\n",
      "250610:1317:52 |INFO| [Cluster: 9] Fetching CellGuide markers...\n",
      "250610:1318:27 |INFO| [Cluster: 8] Annotation not approved, will retry with feedback: Revise the annotation to 'B cell precursor (pro-B or pre-B cell)' based on the dominant expression o...\n",
      "250610:1318:27 |INFO| [Cluster: 8] Retry 2: Re-running complete pipeline with feedback\n",
      "250610:1318:48 |INFO| [Cluster: 7] Annotation not approved, will retry with feedback: Re-examine the top marker gene list for cluster 7. The dominant expression of B cell precursor genes...\n",
      "250610:1318:48 |INFO| [Cluster: 7] Retry 2: Re-running complete pipeline with feedback\n",
      "250610:1318:54 |INFO| [Cluster: 9] Final review with CellGuide markers...\n",
      "250610:1319:37 |INFO| [Cluster: 9] Annotation not approved, will retry with feedback: Revise the annotation to 'early B cell precursor (pro-B/pre-B)' based on the strong and specific exp...\n",
      "250610:1319:37 |INFO| [Cluster: 9] Retry 2: Re-running complete pipeline with feedback\n",
      "250610:1319:38 |INFO| [Cluster: 8] Finding ontology term for annotation...\n",
      "250610:1319:45 |INFO| [Cluster: 7] Finding ontology term for annotation...\n",
      "250610:1319:48 |INFO| [Cluster: 8] Fetching CellGuide markers...\n",
      "250610:1319:48 |INFO| [Cluster: 8] No CellGuide canonical markers found for CL_0000038\n",
      "250610:1319:48 |INFO| [Cluster: 7] Fetching CellGuide markers...\n",
      "250610:1319:53 |INFO| [Cluster: 8] No CellGuide canonical markers found for CL_0000038\n",
      "250610:1320:04 |INFO| [Cluster: 8] No CellGuide canonical markers found for CL_0000038\n",
      "250610:1320:13 |INFO| [Cluster: 9] Finding ontology term for annotation...\n",
      "250610:1320:33 |INFO| [Cluster: 9] Fetching CellGuide markers...\n",
      "250610:1320:38 |INFO| [Cluster: 7] Final review with CellGuide markers...\n",
      "250610:1320:49 |INFO| [Cluster: 8] No CellGuide canonical markers found for CL_0000038\n",
      "250610:1320:49 |INFO| [Cluster: 8] Final review with CellGuide markers...\n",
      "250610:1321:00 |INFO| [Cluster: 7] Reached max retries (2), proceeding with current annotation\n",
      "250610:1321:00 |SUCCESS| [Cluster: 7] Annotation pipeline completed in 352.02s after 2 attempt(s)\n",
      "250610:1321:00 |INFO| [Cluster: 7] LLM Usage: 19 requests, 107159 input tokens, 2567 output tokens, 109726 total tokens\n",
      "250610:1321:00 |INFO| [Cluster: 7] Saving partial result (19/21)\n",
      "250610:1321:16 |INFO| [Cluster: 8] Reached max retries (2), proceeding with current annotation\n",
      "250610:1321:16 |SUCCESS| [Cluster: 8] Annotation pipeline completed in 344.53s after 2 attempt(s)\n",
      "250610:1321:16 |INFO| [Cluster: 8] LLM Usage: 21 requests, 114633 input tokens, 2609 output tokens, 117242 total tokens\n",
      "250610:1321:16 |INFO| [Cluster: 8] Saving partial result (20/21)\n",
      "250610:1321:32 |INFO| [Cluster: 9] Final review with CellGuide markers...\n",
      "250610:1321:50 |INFO| [Cluster: 9] Reached max retries (2), proceeding with current annotation\n",
      "250610:1321:50 |SUCCESS| [Cluster: 9] Annotation pipeline completed in 318.01s after 2 attempt(s)\n",
      "250610:1321:50 |INFO| [Cluster: 9] LLM Usage: 24 requests, 131372 input tokens, 2936 output tokens, 134308 total tokens\n",
      "250610:1321:50 |INFO| [Cluster: 9] Saving partial result (21/21)\n",
      "250610:1321:51 |INFO| [WORKFLOW STEP 3/3]: Generating summary\n",
      "Job 9b2e9900-7410-4380-84ac-ccefcdbe2498 completed successfully.\n",
      "Annotations successfully added to `adata.obs['cytetype_annotation_cell_type']` , ontology term added to `adata.obs['cytetype_cellOntologyTerm_cell_type']` and, full results added to `adata.uns['cytetype_results']`.\n"
     ]
    }
   ],
   "source": [
    "# Roy\n",
    "# Run annotation with custom model\n",
    "adata = annotator.run(\n",
    "    study_context = \"Healthy human hematopoietic tissues from the first trimester (fetal liver), paired second trimester (fetal bone marrow and fetal liver from the same fetuses), pediatric bone marrow, and adult bone marrow, analysed using 10x genomics single cell RNA-seq kits\",\n",
    "    model_config = [{\n",
    "        'provider': 'openai',\n",
    "        'name': 'openai/gpt-4.1',\n",
    "        'apiKey': os.environ.get('OPENROUTER_API_KEY'),\n",
    "        'baseUrl': \"https://openrouter.ai/api/v1\",\n",
    "    }],\n",
    "    auth_token = os.environ.get('NYGEN_API_KEY'),\n",
    "    timeout_seconds=3000,\n",
    "    run_config={\n",
    "        'concurrentClusters': 5,        # Default: 3, Range: 2-10\n",
    "        'maxAnnotationRevisions': 2,    # Default: 2, Range: 1-5\n",
    "        'maxLLMRequests': 525           # Default: 500, Range: 50-2000\n",
    "    },\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc132df6-ad5f-494b-979d-ca962ff7c24b",
   "metadata": {},
   "outputs": [],
   "source": [
    "Author_vs_LLM = pd.crosstab(adata.obs['Level 2 RNA'], adata.obs['cytetype_annotation_Level 2 RNA']).T.idxmax()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e441b7b4-d5b2-4fd8-827a-1bfc1e2720de",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_path = \"/Users/alexantill/Göran_Karlsson_Lab/benchLLM/Zhang_et_al/Qwen3.json\"\n",
    "\n",
    "with open(output_path, \"w\") as f:\n",
    "    json.dump(adata.uns['cytetype_results']['result'], f, indent=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3088241e-c6c2-404a-a45f-8b88d5abe9b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "Author_vs_LLM.reset_index().to_csv(\"/Users/alexantill/Göran_Karlsson_Lab/benchLLM/Zhang_et_al/author_vs_Qwen3.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe018a23-c9f5-48e1-a1cb-4e22c738c7f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "models = [\n",
    "    \"google/gemini-2.5-pro-preview\",\n",
    "    \"anthropic/claude-3.7-sonnet:thinking\",\n",
    "    \"openai/o3-mini\",\n",
    "    \"openai/o4-mini\",\n",
    "    \"x-ai/grok-3-beta\",\n",
    "    \"meta-llama/llama-4-maverick\",\n",
    "    \"anthropic/claude-3.7-sonnet\",\n",
    "    \"openai/gpt-4.1-mini\",\n",
    "    \"qwen/qwen3-235b-a22b\",\n",
    "    \"google/gemini-2.0-flash-001\",\n",
    "    \"openai/gpt-4.1\"\n",
    "]\n",
    "\n",
    "results = {}\n",
    "\n",
    "for model_full in models:\n",
    "    model_name = model_full.split(\"/\")[-1]\n",
    "    adata_result = annotator.run(\n",
    "        study_context=\"Healthy adult human peripheral blood, analysed on a variety of commerical single cell RNA-seq kits\",\n",
    "        model_config=[{\n",
    "            'provider': 'openai',  # update this if provider changes!\n",
    "            'name': model_full,\n",
    "            'apiKey': os.environ.get('OPENROUTER_API_KEY'),\n",
    "            'baseUrl': \"https://openrouter.ai/api/v1\",\n",
    "        }],\n",
    "        auth_token=os.environ.get('NYGEN_API_KEY'),\n",
    "        timeout_seconds=1500,\n",
    "    )\n",
    "    results[f'adata_{model_name}'] = adata_result\n",
    "\n",
    "    # Create crosstab and get idxmax\n",
    "    cross_tab = pd.crosstab(\n",
    "        adata_result.obs.cell_type,\n",
    "        adata_result.obs.cytetype_cell_type\n",
    "    ).T.idxmax()\n",
    "\n",
    "    cross_tab_df = cross_tab.reset_index()\n",
    "    \n",
    "    # Save to CSV with model-specific filename\n",
    "    output_path = f\"/Users/alexantill/Göran_Karlsson_Lab/benchLLM/Simone_et_al/author_vs_{model_name}.csv\"\n",
    "    cross_tab_df.to_csv(output_path, index=False)\n",
    "    print(f\"Saved crosstab for {model_name} to {output_path}\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (benchLLM)",
   "language": "python",
   "name": "benchllm"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
